<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>data_wrangling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.css">

<link rel="stylesheet" href="styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="https://www.vectorbyte.org" class="navbar-brand navbar-brand-logo">
    <img src="./graphics/vblogoedit.pdf" alt="" class="navbar-logo">
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./schedule2023.html">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./materials.html">
 <span class="menu-text">Materials</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/VectorByteOrg/vectorbyte-training4"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/vectorbite_rcn"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#welcome-to-the-vectorbyte-2023-section-on-data-wrangling-and-visualiszation" id="toc-welcome-to-the-vectorbyte-2023-section-on-data-wrangling-and-visualiszation" class="nav-link active" data-scroll-target="#welcome-to-the-vectorbyte-2023-section-on-data-wrangling-and-visualiszation">Welcome to the VectorByte 2023 section on Data Wrangling and Visualiszation</a></li>
  <li><a href="#data-management-and-visualization" id="toc-data-management-and-visualization" class="nav-link" data-scroll-target="#data-management-and-visualization">Data Management and Visualization</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a>
  <ul class="collapse">
  <li><a href="#r-vs.-python" id="toc-r-vs.-python" class="nav-link" data-scroll-target="#r-vs.-python">R vs.&nbsp;Python</a></li>
  </ul></li>
  <li><a href="#data-wrangling" id="toc-data-wrangling" class="nav-link" data-scroll-target="#data-wrangling">Data wrangling</a></li>
  <li><a href="#some-data-wrangling-principles" id="toc-some-data-wrangling-principles" class="nav-link" data-scroll-target="#some-data-wrangling-principles">Some data wrangling principles</a>
  <ul class="collapse">
  <li><a href="#an-example" id="toc-an-example" class="nav-link" data-scroll-target="#an-example">An example</a></li>
  <li><a href="#keep-a-metadata-file-for-each-unique-dataset" id="toc-keep-a-metadata-file-for-each-unique-dataset" class="nav-link" data-scroll-target="#keep-a-metadata-file-for-each-unique-dataset">Keep a metadata file for each unique dataset</a></li>
  <li><a href="#minimize-modifying-raw-data-by-hand" id="toc-minimize-modifying-raw-data-by-hand" class="nav-link" data-scroll-target="#minimize-modifying-raw-data-by-hand">Minimize modifying raw data by hand</a></li>
  <li><a href="#convert-wide-format-data-to-long-format-using-scripts" id="toc-convert-wide-format-data-to-long-format-using-scripts" class="nav-link" data-scroll-target="#convert-wide-format-data-to-long-format-using-scripts">Convert wide format data to long format using scripts</a></li>
  </ul></li>
  <li><a href="#on-to-data-exploration" id="toc-on-to-data-exploration" class="nav-link" data-scroll-target="#on-to-data-exploration">On to data exploration</a>
  <ul class="collapse">
  <li><a href="#and-then-came-tidyverse" id="toc-and-then-came-tidyverse" class="nav-link" data-scroll-target="#and-then-came-tidyverse">And then came <code>tidyverse</code></a></li>
  </ul></li>
  <li><a href="#handling-big-data-in-r" id="toc-handling-big-data-in-r" class="nav-link" data-scroll-target="#handling-big-data-in-r">Handling Big Data in R</a>
  <ul class="collapse">
  <li><a href="#databases-and-r" id="toc-databases-and-r" class="nav-link" data-scroll-target="#databases-and-r">Databases and R</a></li>
  </ul></li>
  <li><a href="#data-visualization" id="toc-data-visualization" class="nav-link" data-scroll-target="#data-visualization">Data visualization</a>
  <ul class="collapse">
  <li><a href="#data-exploration-with-basic-plotting" id="toc-data-exploration-with-basic-plotting" class="nav-link" data-scroll-target="#data-exploration-with-basic-plotting">Data exploration with basic plotting</a></li>
  <li><a href="#scatter-plots" id="toc-scatter-plots" class="nav-link" data-scroll-target="#scatter-plots">Scatter Plots</a></li>
  <li><a href="#histograms" id="toc-histograms" class="nav-link" data-scroll-target="#histograms">Histograms</a></li>
  <li><a href="#subplots" id="toc-subplots" class="nav-link" data-scroll-target="#subplots">Subplots</a></li>
  <li><a href="#overlaying-plots" id="toc-overlaying-plots" class="nav-link" data-scroll-target="#overlaying-plots">Overlaying plots</a></li>
  <li><a href="#boxplots" id="toc-boxplots" class="nav-link" data-scroll-target="#boxplots">Boxplots</a></li>
  <li><a href="#saving-your-graphics" id="toc-saving-your-graphics" class="nav-link" data-scroll-target="#saving-your-graphics">Saving your graphics</a></li>
  </ul></li>
  <li><a href="#beautiful-graphics-in-r" id="toc-beautiful-graphics-in-r" class="nav-link" data-scroll-target="#beautiful-graphics-in-r">Beautiful graphics in R</a>
  <ul class="collapse">
  <li><a href="#quick-plotting-with-qplot" id="toc-quick-plotting-with-qplot" class="nav-link" data-scroll-target="#quick-plotting-with-qplot">Quick plotting with <code>qplot</code></a></li>
  </ul></li>
  <li><a href="#box-plots-in-ggplot" id="toc-box-plots-in-ggplot" class="nav-link" data-scroll-target="#box-plots-in-ggplot">Box plots in ggplot</a>
  <ul class="collapse">
  <li><a href="#some-more-important-ggplot-options" id="toc-some-more-important-ggplot-options" class="nav-link" data-scroll-target="#some-more-important-ggplot-options">Some more important ggplot options</a></li>
  <li><a href="#the-geom-argument" id="toc-the-geom-argument" class="nav-link" data-scroll-target="#the-geom-argument">The <code>geom</code> argument</a></li>
  <li><a href="#advanced-plotting-ggplot" id="toc-advanced-plotting-ggplot" class="nav-link" data-scroll-target="#advanced-plotting-ggplot">Advanced plotting: <code>ggplot</code></a></li>
  </ul></li>
  <li><a href="#mathematical-display" id="toc-mathematical-display" class="nav-link" data-scroll-target="#mathematical-display">Mathematical display</a></li>
  <li><a href="#readings-resources" id="toc-readings-resources" class="nav-link" data-scroll-target="#readings-resources">Readings &amp; Resources</a>
  <ul class="collapse">
  <li><a href="#data-management-reformatting-and-cleaning" id="toc-data-management-reformatting-and-cleaning" class="nav-link" data-scroll-target="#data-management-reformatting-and-cleaning">Data management, reformatting and cleaning</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization">Visualization</a></li>
  </ul></li>
  <li><a href="#experimental-design-and-data-exploration" id="toc-experimental-design-and-data-exploration" class="nav-link" data-scroll-target="#experimental-design-and-data-exploration">Experimental design and Data exploration</a>
  <ul class="collapse">
  <li><a href="#introduction-1" id="toc-introduction-1" class="nav-link" data-scroll-target="#introduction-1">Introduction</a></li>
  <li><a href="#some-statistical-parlance" id="toc-some-statistical-parlance" class="nav-link" data-scroll-target="#some-statistical-parlance">Some statistical parlance</a></li>
  <li><a href="#descriptive-statistics" id="toc-descriptive-statistics" class="nav-link" data-scroll-target="#descriptive-statistics">Descriptive Statistics</a></li>
  <li><a href="#descriptive-statistics-in-r" id="toc-descriptive-statistics-in-r" class="nav-link" data-scroll-target="#descriptive-statistics-in-r">Descriptive statistics in R</a></li>
  <li><a href="#data-types-and-statistical-distributions" id="toc-data-types-and-statistical-distributions" class="nav-link" data-scroll-target="#data-types-and-statistical-distributions">Data types and (statistical) distributions</a></li>
  <li><a href="#sampling-from-distributions-in-r" id="toc-sampling-from-distributions-in-r" class="nav-link" data-scroll-target="#sampling-from-distributions-in-r">Sampling from distributions in R</a></li>
  <li><a href="#two-basic-rules-of-experimental-design-and-sampling" id="toc-two-basic-rules-of-experimental-design-and-sampling" class="nav-link" data-scroll-target="#two-basic-rules-of-experimental-design-and-sampling">Two basic rules of experimental design and sampling</a></li>
  <li><a href="#data-exploration" id="toc-data-exploration" class="nav-link" data-scroll-target="#data-exploration">Data exploration</a>
  <ul class="collapse">
  <li><a href="#some-general-tips" id="toc-some-general-tips" class="nav-link" data-scroll-target="#some-general-tips">Some general tips</a></li>
  <li><a href="#a-case-study" id="toc-a-case-study" class="nav-link" data-scroll-target="#a-case-study">A case study</a></li>
  </ul></li>
  <li><a href="#visualise-distributions-of-the-variables" id="toc-visualise-distributions-of-the-variables" class="nav-link" data-scroll-target="#visualise-distributions-of-the-variables">Visualise distributions of the variables</a></li>
  <li><a href="#take-a-quick-look-at-effects-of-certain-factors" id="toc-take-a-quick-look-at-effects-of-certain-factors" class="nav-link" data-scroll-target="#take-a-quick-look-at-effects-of-certain-factors">Take a quick look at effects of certain factors</a></li>
  <li><a href="#compare-distribution-of-the-variable-across-levels-of-a-factor" id="toc-compare-distribution-of-the-variable-across-levels-of-a-factor" class="nav-link" data-scroll-target="#compare-distribution-of-the-variable-across-levels-of-a-factor">Compare distribution of the variable across levels of a factor</a></li>
  <li><a href="#explore-further-by-scatter-plotting-two-variables" id="toc-explore-further-by-scatter-plotting-two-variables" class="nav-link" data-scroll-target="#explore-further-by-scatter-plotting-two-variables">Explore further by scatter-plotting two variables</a></li>
  <li><a href="#save-your-results" id="toc-save-your-results" class="nav-link" data-scroll-target="#save-your-results">Save your results</a>
  <ul class="collapse">
  <li><a href="#saving-the-exploratory-graphics" id="toc-saving-the-exploratory-graphics" class="nav-link" data-scroll-target="#saving-the-exploratory-graphics">Saving the exploratory graphics</a></li>
  <li><a href="#saving-other-results" id="toc-saving-other-results" class="nav-link" data-scroll-target="#saving-other-results">Saving other results</a></li>
  </ul></li>
  <li><a href="#from-data-exploration-to-statistical-analysis" id="toc-from-data-exploration-to-statistical-analysis" class="nav-link" data-scroll-target="#from-data-exploration-to-statistical-analysis">From data exploration to statistical analysis</a></li>
  <li><a href="#practice-problem-for-data-wrangling" id="toc-practice-problem-for-data-wrangling" class="nav-link" data-scroll-target="#practice-problem-for-data-wrangling">Practice Problem for Data Wrangling</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<p><a href="./materials.html">Main materials</a></p>
<section id="welcome-to-the-vectorbyte-2023-section-on-data-wrangling-and-visualiszation" class="level2">
<h2 class="anchored" data-anchor-id="welcome-to-the-vectorbyte-2023-section-on-data-wrangling-and-visualiszation">Welcome to the VectorByte 2023 section on Data Wrangling and Visualiszation</h2>
<p>This section will cover: * Data Management and Visualization * Experimental Design</p>
</section>
<section id="data-management-and-visualization" class="level2">
<h2 class="anchored" data-anchor-id="data-management-and-visualization">Data Management and Visualization</h2>
<blockquote class="blockquote">
<p><em>Clutter and confusion are failures of design, not attributes of information.</em> – Edward Tuftey</p>
</blockquote>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This section aims at introducing you to key principles and methods for data processing, storage, exploration and visualization.</p>
<p>In this modern world, massive amounts of data are being generated in biology due to rapid advances in technologies for collecting new, as well as for digitizing old data. Some prominent examples are <em>Genomic</em>, <em>Ecosystem respiration</em>, <em>Climatic</em>, and <em>Animal tracking</em> data. Ultimately, the goal of quantitative biology is to both, discover patterns in these data, and fit mathematical models to them. Reproducible data manipulation, analyses and visualization are particularly necessary when data are so large and complex, and few are more so than biological data, which are extremely heterogeneous in their structure and quality. Furthermore, when these data are “big” (more below on what makes a dataset “big”), computationally-efficient data handling, manipulation and analysis techniques are needed.</p>
<section id="r-vs.-python" class="level3">
<h3 class="anchored" data-anchor-id="r-vs.-python">R vs.&nbsp;Python</h3>
<p>We will use R in VectorBiTE 2023 Training because it a great one stop solution for both data manipulation, analysis and visualization. In general, R will do the job for most of your purposes. There is not much between difference these two languages for data science. One advantage that Python has is its greater computational efficiency. On the other hand, R was developed for convenient statistical analyses, with efficiency not being the main concern.</p>
<p>Read more about R vs Python for data science <a href="https://www.guru99.com/r-vs-python.html#:~:text=R%20is%20mainly%20used%20for,of%20course%2C%20the%20ideal%20solution.&amp;text=Python%20is%20a%20general%2Dpurpose%20language%20with%20a%20readable%20syntax.">here</a> and <a href="https://www.datacamp.com/community/tutorials/r-or-python-for-data-analysis">here</a>.</p>
</section>
</section>
<section id="data-wrangling" class="level2">
<h2 class="anchored" data-anchor-id="data-wrangling">Data wrangling</h2>
<p>You are likely to spend far more time than you think dredging through data files manually – checking them, editing them, and reformatting them to make them useful for data exploration and analysis. It is often the case that you’ll have to deal with messy or incomplete data, either because of sampling challenges (e.g., “field” data), or because you got given data that was poorly recorded and maintained. The data we obtain from different data sources is often unusable at the beginning; for example you may need to: * Identify the variables vs observations within the data—somebody else might have recorded the data, or you youself might have collected the data some time back! * Fill in zeros (true measured or observed absences) * Identify and add a value (e.g., <code>-999999</code>) to denote missing observations * Derive or calculate new variables from the raw observations (e.g., convert measurements to SI units; kilograms, meters, seconds, etc.) * Reshape/reformat your data into a layout that works best for analysis (e.g., for R itself);e.g., from wide to long data format for replicated (across plates, chambers, plots, sites, etc.) data * Merge multiple datasets together into a single data sheet</p>
<p>This is not an exhaustive list. Doing so many different things to your raw data is both time-consuming and risky. Why risky? Because to err is very human, and every new, tired mouse-click and/or keyboard-stab has a high probability of inducing an erroneous data point!</p>
</section>
<section id="some-data-wrangling-principles" class="level2">
<h2 class="anchored" data-anchor-id="some-data-wrangling-principles">Some data wrangling principles</h2>
<p>So you would like a record of the data wrangling process (so that it is repeatable and even reversible), and automate it to the extent possible. To this end, here are some guidelines:</p>
<ul>
<li><p>Store data in universally (machine)-readable, non-proprietary formats; basically, use plain ASCII text for your file names, variable/field/column names, and data values. And make sure the data file’s “text encoding” is correct and standard (e.g., <code>UTF-8</code>).</p></li>
<li><p>Keep a metadata file for each unique dataset (again, in non-proprietary format).</p></li>
<li><p>Minimize modifying raw data by hand—use scripts instead—keep a copy of the data as they were recorded.</p></li>
<li><p>Use meaningful names for your data and files and field (column) names</p></li>
<li><p>When you add data, try not to add columns (widening the format); rather, design your tables/data-sheets so that you add only rows (lengthening the format)—and convert “wide format data” to “long format data” using scripts, not by hand,</p></li>
<li><p>All cells within a data column should contain only one type of information (i.e., either text (character), numeric, etc.).</p></li>
<li><p>Ultimately, consider creating a relational database for your data (More on this below).</p></li>
</ul>
<p>This is not an exhaustive list either— see the <a href="#Readings-&amp;-Resources">Readings &amp; Resources Section</a>.</p>
<section id="an-example" class="level3">
<h3 class="anchored" data-anchor-id="an-example">An example</h3>
<p>We will use the Pound Hill dataset collected by students in an Imperial College Field Course for understanding some of these principles. This is not vectorbite data but it is an excellent example of a dataset that needs wrangling. VectorBiTE is too well structured for this sort of wrangling.</p>
<p>To start with, we need to import the <em>raw</em> data file, for which, follow these steps:</p>
<p>★ Copy the file <code>PoundHillData.csv</code> and <code>PoundHillMetaData.csv</code> files from the VectorBiTE <code>data</code> directory into your own R <code>data</code> directory. Then load the data in R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>MyData <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">read.csv</span>(<span class="st">"activities/data/PoundHillData.csv"</span>,<span class="at">header =</span> <span class="cn">FALSE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(MyData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "matrix" "array" </code></pre>
</div>
</div>
<ul>
<li>Loading the data <code>as.matrix()</code>, and setting <code>header=FALSE</code> guarantees that the data are imported “as is” so that you can wrangle them. Otherwise <code>read.csv</code> will convert the first row to column headers.</li>
<li>All the data will be converted to the character class in the resulting matrix called <code>MyData</code> because at least one of the entries is already character class.</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>As of R version 4.0.0 of R released in April 2020, the default for <code>stringsAsFactors</code> <a href="https://cran.r-project.org/doc/manuals/r-devel/NEWS.html">was changed to false</a> If you are using R version 3.x.x, you will need to add <code>stringsAsFactors = FALSE</code> to the above command to prevent R from converting all columns of character type (strings) to the <code>factor</code> data type (this will create problems with subsequent data wrangling).</p>
</div>
</div>
</div>
<p>★ Now load the Metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>MyMetaData <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"activities/data/PoundHillMetaData.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>,  <span class="at">sep=</span><span class="st">";"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(MyMetaData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<ul>
<li>Here, <code>header =TRUE</code> because we do have metadata headers (<code>FieldName</code> and<code>Description</code>), and</li>
<li>We have used semicolon (<code>;</code>) as delimiter because there are commas in one of the field descriptions.</li>
<li>We have avoided spaces in the columns headers (so “FieldName” instead of “Field Name”) — please avoid spaces in field or column names because R will replace each space in a column header with a dot, which may be confusing.</li>
</ul>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip! The text encoding of your data file
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you have a string (<code>character</code> data type) stored as a variable in your R workspace, or a file containing strings, the computer has to know what <a href="https://en.wikipedia.org/wiki/Character_encoding">character encoding</a> it is in or it cannot interpret or display it to you correctly. Usually, the encoding will be UTF-8 or ASCII, which is easily handled by most computer languages. Sometimes you may run into (unexpected) bugs when importing and running scripts in <code>R</code> because your file has a non-standard text encoding. You will need to specify the encoding in that case, using the <code>encoding</code> argument of <code>read.csv()</code> and <code>read.table()</code>. You can check the encoding of a file by using <code>find</code> in Linux/Mac. Try in your UNIX terminal:</p>
<p><code>file -i data/PoundHillData.csv</code></p>
<p>or, check encoding of all files in the <code>Data</code> directory:</p>
<p><code>file -i data/*.csv</code></p>
<p>Use <code>file -I</code> instead of <code>file -i</code> in a Mac terminal.</p>
</div>
</div>
</div>
<p>Now check out what the data look like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(MyData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     V1                     V2        V3        V4        V5        V6       
[1,] "Cultivation"          "october" "october" "october" "october" "october"
[2,] "Block"                "a"       "a"       "a"       "a"       "a"      
[3,] "Plot"                 "1"       "1"       "1"       "1"       "1"      
[4,] "Quadrat"              "Q1"      "Q2"      "Q3"      "Q4"      "Q5"     
[5,] "Achillea millefolium" "4"       "8"       "3"       "20"      "6"      
[6,] "Agrostis gigantea"    ""        ""        ""        ""        "15"     
     V7    V8    V9    V10   V11   V12     V13     V14     V15     V16    
[1,] "may" "may" "may" "may" "may" "march" "march" "march" "march" "march"
[2,] "a"   "a"   "a"   "a"   "a"   "a"     "a"     "a"     "a"     "a"    
[3,] "2"   "2"   "2"   "2"   "2"   "3"     "3"     "3"     "3"     "3"    
[4,] "Q1"  "Q2"  "Q3"  "Q4"  "Q5"  "Q1"    "Q2"    "Q3"    "Q4"    "Q5"   
[5,] "4"   ""    ""    ""    ""    "1"     "2"     "5"     "12"    "2"    
[6,] ""    ""    ""    ""    "2"   "8"     ""      "53"    "16"    "47"   
     V17     V18   V19   V20   V21   V22   V23     V24     V25     V26    
[1,] "march" "may" "may" "may" "may" "may" "march" "march" "march" "march"
[2,] "a"     "b"   "b"   "b"   "b"   "b"   "b"     "b"     "b"     "b"    
[3,] "3"     "4"   "4"   "4"   "4"   "4"   "5"     "5"     "5"     "5"    
[4,] "Q6"    "Q1"  "Q2"  "Q3"  "Q4"  "Q5"  "Q1"    "Q2"    "Q3"    "Q4"   
[5,] ""      ""    ""    ""    ""    ""    "3"     "8"     "3"     "7"    
[6,] "12"    "68"  "57"  "83"  "88"  "89"  "10"    "29"    "24"    "24"   
     V27     V28       V29       V30       V31       V32     V33     V34    
[1,] "march" "october" "october" "october" "october" "march" "march" "march"
[2,] "b"     "b"       "b"       "b"       "b"       "c"     "c"     "c"    
[3,] "5"     "6"       "6"       "6"       "6"       "7"     "7"     "7"    
[4,] "Q5"    "Q1"      "Q2"      "Q3"      "Q4"      "Q1"    "Q2"    "Q3"   
[5,] "20"    "1"       ""        "3"       "1"       "24"    "26"    "37"   
[6,] "43"    "143"     "39"      "10"      "67"      "62"    "22"    "37"   
     V35     V36     V37       V38       V39       V40       V41       V42  
[1,] "march" "march" "october" "october" "october" "october" "october" "may"
[2,] "c"     "c"     "c"       "c"       "c"       "c"       "c"       "c"  
[3,] "7"     "7"     "8"       "8"       "8"       "8"       "8"       "9"  
[4,] "Q4"    "Q5"    "Q1"      "Q2"      "Q3"      "Q4"      "Q5"      "Q1" 
[5,] "37"    "100"   "2"       ""        ""        "1"       ""        ""   
[6,] "4"     "27"    "198"     "96"      "99"      "48"      "351"     "57" 
     V43   V44   V45   V46   V47     V48     V49     V50     V51     V52    
[1,] "may" "may" "may" "may" "march" "march" "march" "march" "march" "march"
[2,] "c"   "c"   "c"   "c"   "d"     "d"     "d"     "d"     "d"     "d"    
[3,] "9"   "9"   "9"   "9"   "10"    "10"    "10"    "10"    "10"    "10"   
[4,] "Q2"  "Q3"  "Q4"  "Q5"  "Q1"    "Q2"    "Q3"    "Q4"    "Q5"    "Q6"   
[5,] ""    "7"   ""    "5"   "7"     "2"     "2"     "9"     "4"     "10"   
[6,] "41"  "65"  "38"  "45"  "259"   "33"    "96"    "291"   "19"    "80"   
     V53   V54   V55   V56       V57       V58       V59       V60      
[1,] "may" "may" "may" "october" "october" "october" "october" "october"
[2,] "d"   "d"   "d"   "d"       "d"       "d"       "d"       "d"      
[3,] "12"  "12"  "12"  "11"      "11"      "11"      "11"      "11"     
[4,] "Q1"  "Q2"  "Q4"  "Q1"      "Q2"      "Q3"      "Q4"      "Q5"     
[5,] ""    ""    ""    ""        ""        "12"      "6"       "5"      
[6,] "33"  "145" "45"  "62"      "25"      "57"      "113"     "12"     </code></pre>
</div>
</div>
<p>Note that column names V1-V60 were generated automatically by R when you imported the data</p>
<p>In RStudio you can also do <code>view(MyData)</code> at the R prompt or any other code editor, <code>fix(MyData)</code>. We won’t do anything with the metadata file in this session except inspect the information it contains.</p>
</section>
<section id="keep-a-metadata-file-for-each-unique-dataset" class="level3">
<h3 class="anchored" data-anchor-id="keep-a-metadata-file-for-each-unique-dataset">Keep a metadata file for each unique dataset</h3>
<p>Data wrangling really begins immediately after data collection. You may collect data of different kinds (e.g., diversity, biomass, tree girth), etc. Keep the original spreadsheet well documented using a “metadata” file that describes the data (you would hopefully have written the first version of this even before you started collecting the data!). The minimum information needed to make a metadata file useful is a description of each of the <em>fields</em> — the column or row headers under which the information is stored in your data/spreadsheet.</p>
<p>Have a look at the metadata file for the Pound Hill dataset:</p>
<p><strong>Boolean arguments in R</strong>: In <code>R</code>, you can use <code>F</code> and <code>T</code> for boolean <code>FALSE</code> and <code>TRUE</code> respectively. To see this, type <code>a &lt;- T</code></p>
<p>in the R commandline, and then see what R returns when you type <code>a</code>. Using <code>F</code> and <code>T</code> for boolean <code>FALSE</code> and <code>TRUE</code> respectively is not necessarily good practice, but be aware that this option exists.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>MyMetaData</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     FieldName
1  Cultivation
2        Block
3         Plot
4      Quadrat
5 SpeciesCount
                                                           Description
1  Cultivation treatments applied in three months: october, may, march
2                                            Treatment blocks ids: a-d
3                                 Plot ids under each treatment : 1-12
4                  Sampling quadrats (25x50 cm each) per plot: Q1--Q6 
5                 Number of individuals of species (count) per quadrat</code></pre>
</div>
</div>
<p>Ideally, you would also like to add more information about the data, such as the measurement units of each type of observation. These data include just one type of observation: Number of individuals of species per sample (plot), which is a count (integer, or <code>int</code> data class).</p>
</section>
<section id="minimize-modifying-raw-data-by-hand" class="level3">
<h3 class="anchored" data-anchor-id="minimize-modifying-raw-data-by-hand">Minimize modifying raw data by hand</h3>
<p>When the dataset is large (e.g., 1000’s of rows), cleaning and exploring it can get tricky, and you are very likely to make many mistakes. You should record all the steps you used to process it with an R script rather than risking a manual and basically <em>irreproducible</em> processing. Most importantly, <em>avoid or minimize editing your raw data file</em>—make a copy (with a meaningful tag in the file name to indicate the date and author) before making hand edits.</p>
<p>All blank cells in the data are <em>true absences</em>, in the sense that species was actually not present in that quadrat. So we can replace those blanks with zeros:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>MyData[MyData <span class="sc">==</span> <span class="st">""</span>] <span class="ot">=</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="convert-wide-format-data-to-long-format-using-scripts" class="level3">
<h3 class="anchored" data-anchor-id="convert-wide-format-data-to-long-format-using-scripts">Convert wide format data to long format using scripts</h3>
<p>One typically records data in the field or experiments using a “wide” format, where a subject’s (e.g., habitat, plot, treatment, species etc) repeated responses or observations (e.g., species count, biomass, etc) will be in a single row, and each response in a separate column. The raw Pound Hill data were recorded in this way. However, the wide format is not ideal for data analysis — instead you need the data in a “long” format, where each row is one observation point per subject. So each subject will have data in multiple rows. Any measures/variables/observations that don’t change across the subjects will have the same value in all the rows. For humans, the wide format is generally more intuitive for recording (e.g., in field data sheets) data. However, for data inspection and analysis, the long format is preferable for two main reasons:</p>
<ul>
<li><p>If you have many response and/or treatment variables, it is hard to inspect the data values in a wide-form version of the dataset. In the case of the pound hill dataset, the response variable is species, and treatment variable is cultivation month (with sequentially nested replicates—block, plot, quadrat— within it), and the data values are the number (count) of individuals of each species per quadrat. As you can see, there are a a large number of columns (60 to be exact), with columns V2-V60 containing different treatment combinations. This makes it hard to visually see problems with the data values. You would have to look across all these columns to see any issues, or if you wanted to run a single command on all the data values (e.g., <code>is.integer()</code> to check if they are all integers, as you would expect), it would be harder to do so or interpret the output if all the species counts were in a single column.</p></li>
<li><p>Long-form datasets are typically required for statistical analysis and visualization packages or commands in R (or Python, for that matter). For example, if you wanted to fit a linear model using R’s <code>lm()</code> command, with treatment (cultivation month) as the independent variable, you would need to structure your data in long form. Similarly, and if you wanted to plot histograms of species numbers by treatment using ggplot (coming up), you would also need these data in long format.</p></li>
</ul>
<p><em>OK, so let’s go from wide to long format already!</em></p>
<p>You can switch between wide and long formats using <code>melt()</code> and <code>dcast()</code> from the <code>reshape2</code> package, as illustrated in the script <code>DataWrang.R</code> available at <a href="https://github.com/mhasoba/TheMulQuaBio/tree/master/content/code">TheMulQuaBio repository</a>. But first, let’s transpose the data, because for a long format, the (nested) treatments variables should be in rows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>MyData <span class="ot">&lt;-</span> <span class="fu">t</span>(MyData) </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(MyData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [,1]          [,2]    [,3]   [,4]      [,5]                  
V1 "Cultivation" "Block" "Plot" "Quadrat" "Achillea millefolium"
V2 "october"     "a"     "1"    "Q1"      "4"                   
V3 "october"     "a"     "1"    "Q2"      "8"                   
V4 "october"     "a"     "1"    "Q3"      "3"                   
V5 "october"     "a"     "1"    "Q4"      "20"                  
V6 "october"     "a"     "1"    "Q5"      "6"                   
   [,6]                [,7]                 [,8]              
V1 "Agrostis gigantea" "Anagallis arvensis" "Anchusa arvensis"
V2 "0"                 "0"                  "0"               
V3 "0"                 "0"                  "3"               
V4 "0"                 "0"                  "1"               
V5 "0"                 "0"                  "1"               
V6 "15"                "0"                  "0"               
   [,9]                 [,10]               [,11]               
V1 "Anisantha sterilis" "Aphanes australis" "Artemesia vulgaris"
V2 "0"                  "0"                 "0"                 
V3 "0"                  "0"                 "0"                 
V4 "0"                  "0"                 "2"                 
V5 "0"                  "0"                 "0"                 
V6 "5"                  "0"                 "0"                 
   [,12]              [,13]                  [,14]              
V1 "Bromus hordaceus" "Cerastium glomeratun" "Chenopodium album"
V2 "0"                "0"                    "0"                
V3 "0"                "2"                    "0"                
V4 "0"                "1"                    "0"                
V5 "0"                "0"                    "0"                
V6 "0"                "0"                    "0"                
   [,15]             [,16]               [,17]               [,18]           
V1 "Cirsium arvense" "Conyza canadensis" "Crepis capillaris" "Crepis paludos"
V2 "0"               "0"                 "0"                 "0"             
V3 "5"               "0"                 "0"                 "0"             
V4 "0"               "0"                 "0"                 "0"             
V5 "0"               "0"                 "0"                 "0"             
V6 "0"               "0"                 "0"                 "0"             
   [,19]              [,20]                [,21]               
V1 "Elytrigia repens" "Epilobium ciliatum" "Erodium cicutarium"
V2 "0"                "1"                  "0"                 
V3 "0"                "0"                  "0"                 
V4 "0"                "0"                  "0"                 
V5 "0"                "0"                  "0"                 
V6 "0"                "0"                  "0"                 
   [,22]                  [,23]           [,24]            [,25]          
V1 "Fallopia convolvulus" "Festuca rubra" "Geranium molle" "Holcus mollis"
V2 "0"                    "0"             "0"              "159"          
V3 "0"                    "0"             "0"              "81"           
V4 "0"                    "0"             "0"              "144"          
V5 "0"                    "0"             "0"              "124"          
V6 "0"                    "0"             "0"              "115"          
   [,26]                  [,27]              [,28]               
V1 "Hypochaeris radicata" "Lactuca serriola" "Medicago lupulina "
V2 "0"                    "0"                "0"                 
V3 "0"                    "0"                "0"                 
V4 "0"                    "0"                "0"                 
V5 "0"                    "0"                "0"                 
V6 "0"                    "0"                "0"                 
   [,29]               [,30]                  [,31]            
V1 "Myosotis arvensis" "Plantago lanceolata " "Polpaver dubium"
V2 "0"                 "0"                    "0"              
V3 "0"                 "0"                    "0"              
V4 "0"                 "0"                    "0"              
V5 "0"                 "0"                    "0"              
V6 "0"                 "0"                    "0"              
   [,32]                    [,33]              [,34]          
V1 "Raphanus raphanistrum " "Rumex acetosella" "Rumex crispus"
V2 "0"                      "25"               "0"            
V3 "0"                      "0"                "0"            
V4 "0"                      "0"                "0"            
V5 "0"                      "6"                "0"            
V6 "0"                      "1"                "0"            
   [,35]               [,36]              [,37]           [,38]              
V1 "Rumex obtusifolia" "Semecio jacobaea" "Sonchus asper" "Spergula arvensis"
V2 "0"                 "0"                "0"             "0"                
V3 "0"                 "0"                "0"             "0"                
V4 "0"                 "0"                "0"             "0"                
V5 "0"                 "0"                "0"             "0"                
V6 "0"                 "0"                "0"             "0"                
   [,39]                [,40]                  [,41]                       
V1 "Stellaria graminea" "Taraxacum officinale" "Tripleurospermum inodorum "
V2 "8"                  "0"                    "11"                        
V3 "4"                  "0"                    "1"                         
V4 "1"                  "0"                    "0"                         
V5 "0"                  "0"                    "0"                         
V6 "1"                  "10"                   "0"                         
   [,42]               [,43]              [,44]            [,45]           
V1 "Veronica arvensis" "Veronica persica" "Viola arvensis" "Vulpia myuros "
V2 "0"                 "1"                "0"              "0"             
V3 "0"                 "0"                "3"              "0"             
V4 "0"                 "0"                "1"              "0"             
V5 "0"                 "0"                "1"              "0"             
V6 "0"                 "0"                "0"              "7"             </code></pre>
</div>
</div>
<p>At this point, you should note that the first row in the data <code>matrix</code> actually contains the column headers, but these are not encoded by R as being column headers. As far as R is concerned, the first row is just another data row. You can check this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(MyData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL</code></pre>
</div>
</div>
<p>We need to fix this. Also, we would like to create a dataframe now, as this is needed by the R commands we will be using, and ultimately, any visualization and analysis we will be doing. To this end, let’s first create a temporary dataframe with just the data, without the column names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>TempData <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(MyData[<span class="sc">-</span><span class="dv">1</span>,],<span class="at">stringsAsFactors =</span> F)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(TempData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        V1 V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20
V2 october  a  1 Q1  4  0  0  0  0   0   0   0   0   0   0   0   0   0   0   1
V3 october  a  1 Q2  8  0  0  3  0   0   0   0   2   0   5   0   0   0   0   0
V4 october  a  1 Q3  3  0  0  1  0   0   2   0   1   0   0   0   0   0   0   0
V5 october  a  1 Q4 20  0  0  1  0   0   0   0   0   0   0   0   0   0   0   0
V6 october  a  1 Q5  6 15  0  0  5   0   0   0   0   0   0   0   0   0   0   0
V7     may  a  2 Q1  4  0  0  1  0   0   0   0   0   7   1   0   0   0   0   0
   V21 V22 V23 V24 V25 V26 V27 V28 V29 V30 V31 V32 V33 V34 V35 V36 V37 V38 V39
V2   0   0   0   0 159   0   0   0   0   0   0   0  25   0   0   0   0   0   8
V3   0   0   0   0  81   0   0   0   0   0   0   0   0   0   0   0   0   0   4
V4   0   0   0   0 144   0   0   0   0   0   0   0   0   0   0   0   0   0   1
V5   0   0   0   0 124   0   0   0   0   0   0   0   6   0   0   0   0   0   0
V6   0   0   0   0 115   0   0   0   0   0   0   0   1   0   0   0   0   0   1
V7  16  12   0   0 120   0   0   0   0   0   0   0   1   0   0   0   0   2   0
   V40 V41 V42 V43 V44 V45
V2   0  11   0   1   0   0
V3   0   1   0   0   3   0
V4   0   0   0   0   1   0
V5   0   0   0   0   1   0
V6  10   0   0   0   0   7
V7   0   0   0   2   0   0</code></pre>
</div>
</div>
<p>Note that <code>stringsAsFactors = F</code> is important here because we don’t want R to convert columns to the factor class without supervision. This might be a good idea in may cases, but let’s just do it manually later. This will also force you to think carefully about what data each of you columns contains.</p>
<p>Now let’s assign the original column names to the temporary dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TempData) <span class="ot">&lt;-</span> MyData[<span class="dv">1</span>,] <span class="co"># assign column names from original data</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(TempData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Cultivation Block Plot Quadrat Achillea millefolium Agrostis gigantea
V2     october     a    1      Q1                    4                 0
V3     october     a    1      Q2                    8                 0
V4     october     a    1      Q3                    3                 0
V5     october     a    1      Q4                   20                 0
V6     october     a    1      Q5                    6                15
V7         may     a    2      Q1                    4                 0
   Anagallis arvensis Anchusa arvensis Anisantha sterilis Aphanes australis
V2                  0                0                  0                 0
V3                  0                3                  0                 0
V4                  0                1                  0                 0
V5                  0                1                  0                 0
V6                  0                0                  5                 0
V7                  0                1                  0                 0
   Artemesia vulgaris Bromus hordaceus Cerastium glomeratun Chenopodium album
V2                  0                0                    0                 0
V3                  0                0                    2                 0
V4                  2                0                    1                 0
V5                  0                0                    0                 0
V6                  0                0                    0                 0
V7                  0                0                    0                 7
   Cirsium arvense Conyza canadensis Crepis capillaris Crepis paludos
V2               0                 0                 0              0
V3               5                 0                 0              0
V4               0                 0                 0              0
V5               0                 0                 0              0
V6               0                 0                 0              0
V7               1                 0                 0              0
   Elytrigia repens Epilobium ciliatum Erodium cicutarium Fallopia convolvulus
V2                0                  1                  0                    0
V3                0                  0                  0                    0
V4                0                  0                  0                    0
V5                0                  0                  0                    0
V6                0                  0                  0                    0
V7                0                  0                 16                   12
   Festuca rubra Geranium molle Holcus mollis Hypochaeris radicata
V2             0              0           159                    0
V3             0              0            81                    0
V4             0              0           144                    0
V5             0              0           124                    0
V6             0              0           115                    0
V7             0              0           120                    0
   Lactuca serriola Medicago lupulina  Myosotis arvensis Plantago lanceolata 
V2                0                  0                 0                    0
V3                0                  0                 0                    0
V4                0                  0                 0                    0
V5                0                  0                 0                    0
V6                0                  0                 0                    0
V7                0                  0                 0                    0
   Polpaver dubium Raphanus raphanistrum  Rumex acetosella Rumex crispus
V2               0                      0               25             0
V3               0                      0                0             0
V4               0                      0                0             0
V5               0                      0                6             0
V6               0                      0                1             0
V7               0                      0                1             0
   Rumex obtusifolia Semecio jacobaea Sonchus asper Spergula arvensis
V2                 0                0             0                 0
V3                 0                0             0                 0
V4                 0                0             0                 0
V5                 0                0             0                 0
V6                 0                0             0                 0
V7                 0                0             0                 2
   Stellaria graminea Taraxacum officinale Tripleurospermum inodorum 
V2                  8                    0                         11
V3                  4                    0                          1
V4                  1                    0                          0
V5                  0                    0                          0
V6                  1                   10                          0
V7                  0                    0                          0
   Veronica arvensis Veronica persica Viola arvensis Vulpia myuros 
V2                 0                1              0              0
V3                 0                0              3              0
V4                 0                0              1              0
V5                 0                0              1              0
V6                 0                0              0              7
V7                 0                2              0              0</code></pre>
</div>
</div>
<p>The row names still remain, but they are not really bothering us, so we can ignore them. But if you do want to get rid of them, you can:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(TempData) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(TempData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cultivation Block Plot Quadrat Achillea millefolium Agrostis gigantea
1     october     a    1      Q1                    4                 0
2     october     a    1      Q2                    8                 0
3     october     a    1      Q3                    3                 0
4     october     a    1      Q4                   20                 0
5     october     a    1      Q5                    6                15
6         may     a    2      Q1                    4                 0
  Anagallis arvensis Anchusa arvensis Anisantha sterilis Aphanes australis
1                  0                0                  0                 0
2                  0                3                  0                 0
3                  0                1                  0                 0
4                  0                1                  0                 0
5                  0                0                  5                 0
6                  0                1                  0                 0
  Artemesia vulgaris Bromus hordaceus Cerastium glomeratun Chenopodium album
1                  0                0                    0                 0
2                  0                0                    2                 0
3                  2                0                    1                 0
4                  0                0                    0                 0
5                  0                0                    0                 0
6                  0                0                    0                 7
  Cirsium arvense Conyza canadensis Crepis capillaris Crepis paludos
1               0                 0                 0              0
2               5                 0                 0              0
3               0                 0                 0              0
4               0                 0                 0              0
5               0                 0                 0              0
6               1                 0                 0              0
  Elytrigia repens Epilobium ciliatum Erodium cicutarium Fallopia convolvulus
1                0                  1                  0                    0
2                0                  0                  0                    0
3                0                  0                  0                    0
4                0                  0                  0                    0
5                0                  0                  0                    0
6                0                  0                 16                   12
  Festuca rubra Geranium molle Holcus mollis Hypochaeris radicata
1             0              0           159                    0
2             0              0            81                    0
3             0              0           144                    0
4             0              0           124                    0
5             0              0           115                    0
6             0              0           120                    0
  Lactuca serriola Medicago lupulina  Myosotis arvensis Plantago lanceolata 
1                0                  0                 0                    0
2                0                  0                 0                    0
3                0                  0                 0                    0
4                0                  0                 0                    0
5                0                  0                 0                    0
6                0                  0                 0                    0
  Polpaver dubium Raphanus raphanistrum  Rumex acetosella Rumex crispus
1               0                      0               25             0
2               0                      0                0             0
3               0                      0                0             0
4               0                      0                6             0
5               0                      0                1             0
6               0                      0                1             0
  Rumex obtusifolia Semecio jacobaea Sonchus asper Spergula arvensis
1                 0                0             0                 0
2                 0                0             0                 0
3                 0                0             0                 0
4                 0                0             0                 0
5                 0                0             0                 0
6                 0                0             0                 2
  Stellaria graminea Taraxacum officinale Tripleurospermum inodorum 
1                  8                    0                         11
2                  4                    0                          1
3                  1                    0                          0
4                  0                    0                          0
5                  1                   10                          0
6                  0                    0                          0
  Veronica arvensis Veronica persica Viola arvensis Vulpia myuros 
1                 0                1              0              0
2                 0                0              3              0
3                 0                0              1              0
4                 0                0              1              0
5                 0                0              0              7
6                 0                2              0              0</code></pre>
</div>
</div>
<p>Finally, let’s convert the data to long format. For this, we need the <code>reshape2</code> package:</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Both <code>library()</code> and <code>require()</code> are commands/functions to load packages. The difference is that <code>require()</code> is designed for use inside other functions, so it returns <code>FALSE</code> and gives a warning, whereas<code>library()</code> returns an error by default if the package does not exist.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(reshape2)<span class="co"># load the reshape2 package</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: reshape2</code></pre>
</div>
</div>
<p>Check out the help for the <code>melt</code> command of <code>reshape2</code>: <code>?melt</code>.</p>
<p>OK finally, let’s wrangle this dataset into submission!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>MyWrangledData <span class="ot">&lt;-</span> <span class="fu">melt</span>(TempData, <span class="at">id=</span><span class="fu">c</span>(<span class="st">"Cultivation"</span>, <span class="st">"Block"</span>, <span class="st">"Plot"</span>, <span class="st">"Quadrat"</span>), <span class="at">variable.name =</span> <span class="st">"Species"</span>, <span class="at">value.name =</span> <span class="st">"Count"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(MyWrangledData); <span class="fu">tail</span>(MyWrangledData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cultivation Block Plot Quadrat              Species Count
1     october     a    1      Q1 Achillea millefolium     4
2     october     a    1      Q2 Achillea millefolium     8
3     october     a    1      Q3 Achillea millefolium     3
4     october     a    1      Q4 Achillea millefolium    20
5     october     a    1      Q5 Achillea millefolium     6
6         may     a    2      Q1 Achillea millefolium     4</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>     Cultivation Block Plot Quadrat        Species Count
2414         may     d   12      Q4 Vulpia myuros      0
2415     october     d   11      Q1 Vulpia myuros      0
2416     october     d   11      Q2 Vulpia myuros      0
2417     october     d   11      Q3 Vulpia myuros      0
2418     october     d   11      Q4 Vulpia myuros      0
2419     october     d   11      Q5 Vulpia myuros      0</code></pre>
</div>
</div>
<p>This is the long format we have been extolling! It looks much simpler, with all the count data in a single, long column with each count value associated with some “attributes”. It is now clear that the fundamental unit of analysis is a species’ count in a quadrat, each nested within a plot, each in turn nested within a block, and finally, each of which is nested within a cultivation month (the main treatment).</p>
<p>We can also now assign the correct data types to each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>MyWrangledData[, <span class="st">"Cultivation"</span>] <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(MyWrangledData[, <span class="st">"Cultivation"</span>])</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>MyWrangledData[, <span class="st">"Block"</span>] <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(MyWrangledData[, <span class="st">"Block"</span>])</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>MyWrangledData[, <span class="st">"Plot"</span>] <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(MyWrangledData[, <span class="st">"Plot"</span>])</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>MyWrangledData[, <span class="st">"Quadrat"</span>] <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(MyWrangledData[, <span class="st">"Quadrat"</span>])</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>MyWrangledData[, <span class="st">"Count"</span>] <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(MyWrangledData[, <span class="st">"Count"</span>])</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(MyWrangledData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   2419 obs. of  6 variables:
 $ Cultivation: Factor w/ 3 levels "march","may",..: 3 3 3 3 3 2 2 2 2 2 ...
 $ Block      : Factor w/ 4 levels "a","b","c","d": 1 1 1 1 1 1 1 1 1 1 ...
 $ Plot       : Factor w/ 12 levels "1","10","11",..: 1 1 1 1 1 5 5 5 5 5 ...
 $ Quadrat    : Factor w/ 6 levels "Q1","Q2","Q3",..: 1 2 3 4 5 1 2 3 4 5 ...
 $ Species    : Factor w/ 41 levels "Achillea millefolium",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ Count      : int  4 8 3 20 6 4 0 0 0 0 ...</code></pre>
</div>
</div>
</section>
</section>
<section id="on-to-data-exploration" class="level2">
<h2 class="anchored" data-anchor-id="on-to-data-exploration">On to data exploration</h2>
<p>Once you have wrangled the Pound Hill data to its long format, you are ready to go! You may want to start by examining and visualizing the basic properties of the data, such as the number of tree species (41) in the dataset, number of quadrats (replicates) per plot and cultivation treatment, etc.</p>
<p>We will learn about visualization next. After that, you can return to this dataset and try out some visual data exploration. For example, a useful visualization would be to make a histogram of abundances of species, grouped by different factors. For example, you can look at distributions of species’ abundances grouped by the fundamental treatment, <code>Cultivation</code>.</p>
<section id="and-then-came-tidyverse" class="level3">
<h3 class="anchored" data-anchor-id="and-then-came-tidyverse">And then came <code>tidyverse</code></h3>
<p>So if you think this is the end of the options you have for data wrangling in R, think again. There is a whole data science “ecosystem” you can use in R through the <code>tidyverse</code> package. This meta-package includes <code>dplyr</code>, the next iteration of <code>plyr</code> that addresses the speed issues in the latter, and <code>tidyr</code>, essentially a nicer wrapper to the <code>reshape2</code> package with additional functions, and <code>ggplot2</code> (coming up later).</p>
<p>Please install and load the <code>tidyverse</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: tidyverse</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.7     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()</code></pre>
</div>
</div>
<p>This loads a number of packages at one go (and you might get some name-conflict-masking messages; see the note below). You can list these packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tidyverse_packages</span>(<span class="at">include_self =</span> <span class="cn">TRUE</span>) <span class="co"># the include_self = TRUE means list "tidyverse" as well </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "broom"         "cli"           "crayon"        "dbplyr"       
 [5] "dplyr"         "dtplyr"        "forcats"       "googledrive"  
 [9] "googlesheets4" "ggplot2"       "haven"         "hms"          
[13] "httr"          "jsonlite"      "lubridate"     "magrittr"     
[17] "modelr"        "pillar"        "purrr"         "readr"        
[21] "readxl"        "reprex"        "rlang"         "rstudioapi"   
[25] "rvest"         "stringr"       "tibble"        "tidyr"        
[29] "xml2"          "tidyverse"    </code></pre>
</div>
</div>
<p>That’s a lot of packages!</p>
<p>The first time you load tidyverse, you will get some feedback from R, including something like the following:</p>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()</code></pre>
<p>Depending on which package you are loading, you might get such a warning message, that “so and so” function is being “masked” from another one. Basically, any additional packages you load may have one or more functions with the same names(s) as existing function(s) (which will typically do something completely different) in base R’s packages. This creates name conflicts. Therefore, R will prevent you from using the new function(s) that you have enabled by using the same name of an pre-existing function. It gives priority to the pre-existing one. So, for example, if you call just <code>filter()</code>, the command that will be used is the one from <code>stats</code>, and not the one from <code>dplyr</code>. In this scenario, you can use the new function by using the double colon notation <code>::</code> like so: <code>dplyr::filter()</code>.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <code>::</code> notation is like the dot notation in Python — it allows you to access a particular function (like a Python module) from a package. So you can also use the command at the end of the <code>::</code> (try just <code>as_tibble(MyWrangledData)</code>, for example), but it’s just safer to use the <code>::</code> because then you can rest assured that you are actually using the function you intended to.</p>
</div>
</div>
</div>
<p>More of <code>::</code> usage coming up below!</p>
<p>OK, let’s try data exploration in tidyverse.</p>
<p>First, let’s convert the dataframe to a “tibble”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>tibble<span class="sc">::</span><span class="fu">as_tibble</span>(MyWrangledData) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,419 × 6
   Cultivation Block Plot  Quadrat Species              Count
   &lt;fct&gt;       &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;   &lt;fct&gt;                &lt;int&gt;
 1 october     a     1     Q1      Achillea millefolium     4
 2 october     a     1     Q2      Achillea millefolium     8
 3 october     a     1     Q3      Achillea millefolium     3
 4 october     a     1     Q4      Achillea millefolium    20
 5 october     a     1     Q5      Achillea millefolium     6
 6 may         a     2     Q1      Achillea millefolium     4
 7 may         a     2     Q2      Achillea millefolium     0
 8 may         a     2     Q3      Achillea millefolium     0
 9 may         a     2     Q4      Achillea millefolium     0
10 may         a     2     Q5      Achillea millefolium     0
# … with 2,409 more rows</code></pre>
</div>
</div>
<p>OK, there’s the funky double colon (<code>::</code>) notation again…</p>
<p>Also, note that we didn’t have to reassign <code>MyWrangledData</code> back to its original name because tibble just does the needful to the data frame in its current place.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>A “tibble” in tidyverse is equivalent to R’s traditional <code>data.frame</code>. Tibbles <em>are</em> modified data frames, but they make data exploration even easier.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">glimpse</span>(MyWrangledData) <span class="co">#like str(), but nicer!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 2,419
Columns: 6
$ Cultivation &lt;fct&gt; october, october, october, october, october, may, may, may…
$ Block       &lt;fct&gt; a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, b, b, b, b…
$ Plot        &lt;fct&gt; 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4…
$ Quadrat     &lt;fct&gt; Q1, Q2, Q3, Q4, Q5, Q1, Q2, Q3, Q4, Q5, Q1, Q2, Q3, Q4, Q5…
$ Species     &lt;fct&gt; Achillea millefolium, Achillea millefolium, Achillea mille…
$ Count       &lt;int&gt; 4, 8, 3, 20, 6, 4, 0, 0, 0, 0, 1, 2, 5, 12, 2, 0, 0, 0, 0,…</code></pre>
</div>
</div>
<p>Or this:</p>
<p><code>utils::View(MyWrangledData) #same as fix()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(MyWrangledData, Count<span class="sc">&gt;</span><span class="dv">100</span>) <span class="co">#like subset(), but nicer!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Cultivation Block Plot Quadrat            Species Count
1      october     b    6      Q1  Agrostis gigantea   143
2      october     c    8      Q1  Agrostis gigantea   198
3      october     c    8      Q5  Agrostis gigantea   351
4        march     d   10      Q1  Agrostis gigantea   259
5        march     d   10      Q4  Agrostis gigantea   291
6          may     d   12      Q2  Agrostis gigantea   145
7      october     d   11      Q4  Agrostis gigantea   113
8          may     c    9      Q2 Anisantha sterilis   120
9      october     a    1      Q1      Holcus mollis   159
10     october     a    1      Q3      Holcus mollis   144
11     october     a    1      Q4      Holcus mollis   124
12     october     a    1      Q5      Holcus mollis   115
13         may     a    2      Q1      Holcus mollis   120
14     october     b    6      Q2      Holcus mollis   129
15     october     b    6      Q3      Holcus mollis   123
16     october     c    8      Q2      Holcus mollis   256
17     october     c    8      Q3      Holcus mollis   147
18     october     c    8      Q4      Holcus mollis   128
19     october     d   11      Q1      Holcus mollis   104
20     october     d   11      Q2      Holcus mollis   131
21       march     d   10      Q3 Medicago lupulina    114</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">slice</span>(MyWrangledData, <span class="dv">10</span><span class="sc">:</span><span class="dv">15</span>) <span class="co"># Look at an arbitrary set of data rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cultivation Block Plot Quadrat              Species Count
1         may     a    2      Q5 Achillea millefolium     0
2       march     a    3      Q1 Achillea millefolium     1
3       march     a    3      Q2 Achillea millefolium     2
4       march     a    3      Q3 Achillea millefolium     5
5       march     a    3      Q4 Achillea millefolium    12
6       march     a    3      Q5 Achillea millefolium     2</code></pre>
</div>
</div>
<p>Learning to use <code>tidyverse</code> involves learning some new syntax and a lot of new commands, but if you plan to do a lot of data wrangling and exploration, you will benefit from getting to know them well.</p>
<p>Have a look at some tidyverse-related <a href="https://rstudio.com/resources/cheatsheets/">cheatsheets</a> to explore further.</p>
<p>One additional thing that you would need to get used to in the <code>tidyverse</code> is piping multiple commands using the <code>%%</code> notation, which makes your code ore compact, and in theory, more readable.</p>
</section>
</section>
<section id="handling-big-data-in-r" class="level2">
<h2 class="anchored" data-anchor-id="handling-big-data-in-r">Handling Big Data in R</h2>
<p>The buzzword ‘Big Data’ basically refers to datasets that have the following properties:</p>
<ol type="1">
<li><p>A dataset that does not fit into available RAM on one system (say, 2 gigabytes).</p></li>
<li><p>A dataset that has so many rows (when in it’s long format — see above sections) that it <em>significantly</em> slows down your analysis or simulation without vectorization (that is, when looping).</p></li>
</ol>
<p>Both these criteria are programming language-, and computer hardware-dependent, of course. For example, a 32-bit OS can only handle ~2 GB of RAM, so this computer will struggle every time you try to handle a dataset in that size range.</p>
<p>R reads data into RAM all at once when you using the <code>read.table</code> (or its wrapper, <code>read.csv()</code> — maybe you have realized by now that <code>read.csv()</code> is basically calling <code>read.table()</code> with a particular set of options. That is, objects in R live in memory entirely, and big-ish data in RAM will cause R to choke.</p>
<p>There are a few options (which you can combine, of course) if you are actually using datasets that are so large:</p>
<ul>
<li><p>Import large files smartly; e.g., using <code>scan()</code> in R, and then create subsets of the data that you need. Also, use the <code>reshape</code> or <code>tidyr</code> packages to covert your data in the most “square” (so neither too long or too wide) format as possible. Of course, you will need subsets of data in long format for analysis (see sections above).</p></li>
<li><p>Use the <code>bigmemory</code> package to load data in the gb range (e.g., use <code>read.big.matrix()</code> instead of <code>read.table()</code>. This package also has other useful functions, such as <code>foreach()</code> instead of <code>for()</code> for better memory management.</p></li>
<li><p>Use a 64 bit version of R with enough memory and preferably on UNIX/Linux!</p></li>
<li><p>Vectorize your analyses/simulations to the extent possible.</p></li>
<li><p>Use databases (more on this below).</p></li>
<li><p>Use distributed computing (distribute the analysis/simulation across multiple CPU’s).</p></li>
</ul>
<section id="databases-and-r" class="level3">
<h3 class="anchored" data-anchor-id="databases-and-r">Databases and R</h3>
<p>R can be used to link to and extract data from online databases such as PubMed and GenBank, or to manipulate and access your own. Computational Biology datasets are often quite large, and it makes sense to access their data by querying the databases instead of manually downloading them. So also, your own data may be complex and large, in which case you may want to organize and manage those data in a proper relational database.</p>
<p>Practically all the data wrangling principles in the previous sections are a part and parcel of relational databases.</p>
<p>There are many R packages that provide an interface to databases (SQLite, MySQL, Oracle, etc). Check out R packages <a href="http://cran.r-project.org/web/packages/DBI/index.html">DBI</a> and <a href="https://cran.r-project.org/web/packages/RMySQL/index.html">RMySQL</a> .</p>
</section>
</section>
<section id="data-visualization" class="level2">
<h2 class="anchored" data-anchor-id="data-visualization">Data visualization</h2>
<p>Now that you have learned how to wrangle data, let’s learn some key principles of graphics and visualization, and how to implement them in R. You can use R to build a complete, reproducible workflow for data visualization for both exploratory and publication purposes. We will start with some basic plotting and data exploration. You will then learn to generate publication-quality visualizations using the <code>ggplot2</code> package.</p>
<section id="data-exploration-with-basic-plotting" class="level3">
<h3 class="anchored" data-anchor-id="data-exploration-with-basic-plotting">Data exploration with basic plotting</h3>
<p>Before you do any statistical analyses with data, you must clean, explore, and visualize it. And eventually, you want to produce a finished product that presents visualizations of your data and your results clearly and concisely.</p>
<p>Ultimately, at both, the data exploration and the finished product stages, the goal of graphics is to present information such that it provides intuitive ideas. As Edward Tufte says: <em>“Graphical excellence is that which gives to the viewer the greatest number of ideas in the shortest time with the least ink in the smallest space.”</em></p>
<p>R can produce beautiful graphics without the time-consuming and fiddly methods that you might have used in Excel or equivalent. You should also make it a habit to quickly plot the data for exploratory analysis. So we are going to learn some basic plotting first.</p>
<section id="basic-plotting-commands-in-r" class="level4">
<h4 class="anchored" data-anchor-id="basic-plotting-commands-in-r">Basic plotting commands in R</h4>
<p>Here is a menu of basic R plotting commands (use <code>?commandname</code> to learn more about it):</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Command</th>
<th style="text-align: left;">What it does</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>plot(x,y)</code></td>
<td style="text-align: left;">Scatterplot</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>plot(y~x)</code></td>
<td style="text-align: left;">Scatterplot with <code>y</code> as a response variable</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>hist(mydata)</code></td>
<td style="text-align: left;">Histogram</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>barplot(mydata)</code></td>
<td style="text-align: left;">Bar plot</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>points(y1$\sim$x1)</code></td>
<td style="text-align: left;">Add another series of points</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>boxplot(y$\sim$x)</code></td>
<td style="text-align: left;">Boxplot</td>
</tr>
</tbody>
</table>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>R graphics devices</strong>: In all that follows, you may often end up plotting multiple plots on the same graphics window without intending to do so, because R by default keeps plotting in the most recent plotting window that was opened. You can close a particular graphics window or “device” by using <code>dev.off()</code>, and all open devices/windows with <code>graphics.off()</code>. By default, <code>dev.off()</code> will close the most recent figure device that was opened.</p>
</div>
</div>
</div>
<p>Note that there are invisible devices as well! For example, if you are printing to pdf (coming up below), the device or graphics window will not be visible on your computer screen.</p>
<p>Now let’s try some simple plotting for data exploration. We will use a dataset from a recent paper by our team at Imperial College London. In this paper (<a href="https://royalsocietypublishing.org/doi/10.1098/rspb.2020.3217"><strong>Huxley et al.&nbsp;2021</strong></a>), we investigated how temperature and resource supply can affect an important vector’s life history traits (development, survival, size, fecundity) and its maximal population growth rate.</p>
<p><span class="math inline">\star</span> In <a href="https://vectorbyte.crc.nd.edu/vectraits-explorer">VecTraits</a>, locate and download dataset #558 into your own data directory. Then, launch R and read in these data to a dataframe (note the relative path):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>MyDF <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'activities/data/VByte_558_longevity.csv'</span>,<span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(MyDF) <span class="co">#check the size of the data frame you loaded</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 156  42</code></pre>
</div>
</div>
<p>Let’s look at what the data contain. Type <code>MyDF</code> and hit the TAB key twice in the R command line.</p>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>##Remember Tabbing:In your smart code editor or IDE such as vscode or RStudio, you will see a drop-down list of all the column headers in a dataframe when you hit TAB once after <code>MyDF$</code></p>
</div>
</div>
</div>
<p>You can also use <code>str()</code> and <code>head()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   156 obs. of  42 variables:
 $ Id                       : int  92846 92848 92850 92852 92854 92856 92858 92860 92862 92864 ...
 $ DatasetID                : int  558 558 558 558 558 558 558 558 558 558 ...
 $ IndividualID             : Factor w/ 156 levels "PHX331","PHX332",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ OriginalTraitName        : Factor w/ 1 level "longevity": 1 1 1 1 1 1 1 1 1 1 ...
 $ OriginalTraitDef         : Factor w/ 1 level "individual-level duration of life stage": 1 1 1 1 1 1 1 1 1 1 ...
 $ OriginalTraitValue       : num  8 5 5 6 4 6 9 1 7 7 ...
 $ OriginalTraitUnit        : Factor w/ 1 level "days": 1 1 1 1 1 1 1 1 1 1 ...
 $ Habitat                  : Factor w/ 1 level "terrestrial": 1 1 1 1 1 1 1 1 1 1 ...
 $ LabField                 : Factor w/ 1 level "laboratory": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressor           : Factor w/ 1 level "resource supply": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressorDef        : Factor w/ 1 level "food quantity supplied at a constant rate ": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressorValue      : num  0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 ...
 $ SecondStressorUnit       : Factor w/ 1 level "mg individual-1 day-1": 1 1 1 1 1 1 1 1 1 1 ...
 $ LocationText             : Factor w/ 1 level "Fort Myers Florida": 1 1 1 1 1 1 1 1 1 1 ...
 $ LocationType             : Factor w/ 1 level "field": 1 1 1 1 1 1 1 1 1 1 ...
 $ CoordinateType           : Factor w/ 1 level "decimal": 1 1 1 1 1 1 1 1 1 1 ...
 $ Latitude                 : num  26.6 26.6 26.6 26.6 26.6 ...
 $ Longitude                : num  -81.8 -81.8 -81.8 -81.8 -81.8 ...
 $ Interactor1              : Factor w/ 1 level "Aedes aegypti": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Common        : Factor w/ 1 level "Yellow fever mosquito": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Kingdom       : Factor w/ 1 level "Animalia": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Phylum        : Factor w/ 1 level "Arthropoda": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Class         : Factor w/ 1 level "Insecta": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Order         : Factor w/ 1 level "Diptera": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Family        : Factor w/ 1 level "Culicidae": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Genus         : Factor w/ 1 level "Aedes": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Species       : Factor w/ 1 level "aegypti": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Stage         : Factor w/ 1 level "adult (starved)": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Sex           : Factor w/ 1 level "female": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Temp          : num  22 22 22 22 22 22 22 22 22 22 ...
 $ Interactor1TempUnit      : Factor w/ 1 level "Celsius": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1GrowthTemp    : num  26 26 26 26 26 26 26 26 26 26 ...
 $ Interactor1GrowthTempUnit: Factor w/ 1 level "Celsius": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Size          : num  2.93 2.59 2.41 2.62 2.28 2.39 2.99 2.5 2.75 2.82 ...
 $ Interactor1SizeUnit      : Factor w/ 1 level "mm": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1SizeType      : Factor w/ 1 level "wing length": 1 1 1 1 1 1 1 1 1 1 ...
 $ FigureTable              : Factor w/ 1 level "supplementary material": 1 1 1 1 1 1 1 1 1 1 ...
 $ Citation                 : Factor w/ 1 level "Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. "| __truncated__: 1 1 1 1 1 1 1 1 1 1 ...
 $ DOI                      : Factor w/ 1 level "10.1098/rspb.2020.3217": 1 1 1 1 1 1 1 1 1 1 ...
 $ SubmittedBy              : Factor w/ 1 level "Paul Huxley": 1 1 1 1 1 1 1 1 1 1 ...
 $ ContributorEmail         : Factor w/ 1 level "phuxly@gmail.com": 1 1 1 1 1 1 1 1 1 1 ...
 $ Notes                    : Factor w/ 2 levels "","wing length was not measured for this individual: zero assigned ": 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Id DatasetID IndividualID OriginalTraitName
1 92846       558       PHX331         longevity
2 92848       558       PHX332         longevity
3 92850       558       PHX333         longevity
4 92852       558       PHX334         longevity
5 92854       558       PHX335         longevity
6 92856       558       PHX336         longevity
                         OriginalTraitDef OriginalTraitValue OriginalTraitUnit
1 individual-level duration of life stage                  8              days
2 individual-level duration of life stage                  5              days
3 individual-level duration of life stage                  5              days
4 individual-level duration of life stage                  6              days
5 individual-level duration of life stage                  4              days
6 individual-level duration of life stage                  6              days
      Habitat   LabField  SecondStressor
1 terrestrial laboratory resource supply
2 terrestrial laboratory resource supply
3 terrestrial laboratory resource supply
4 terrestrial laboratory resource supply
5 terrestrial laboratory resource supply
6 terrestrial laboratory resource supply
                           SecondStressorDef SecondStressorValue
1 food quantity supplied at a constant rate                  0.1
2 food quantity supplied at a constant rate                  0.1
3 food quantity supplied at a constant rate                  0.1
4 food quantity supplied at a constant rate                  0.1
5 food quantity supplied at a constant rate                  0.1
6 food quantity supplied at a constant rate                  0.1
     SecondStressorUnit       LocationText LocationType CoordinateType Latitude
1 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
2 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
3 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
4 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
5 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
6 mg individual-1 day-1 Fort Myers Florida        field        decimal 26.61667
  Longitude   Interactor1     Interactor1Common Interactor1Kingdom
1 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
2 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
3 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
4 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
5 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
6 -81.83333 Aedes aegypti Yellow fever mosquito           Animalia
  Interactor1Phylum Interactor1Class Interactor1Order Interactor1Family
1        Arthropoda          Insecta          Diptera         Culicidae
2        Arthropoda          Insecta          Diptera         Culicidae
3        Arthropoda          Insecta          Diptera         Culicidae
4        Arthropoda          Insecta          Diptera         Culicidae
5        Arthropoda          Insecta          Diptera         Culicidae
6        Arthropoda          Insecta          Diptera         Culicidae
  Interactor1Genus Interactor1Species Interactor1Stage Interactor1Sex
1            Aedes            aegypti  adult (starved)         female
2            Aedes            aegypti  adult (starved)         female
3            Aedes            aegypti  adult (starved)         female
4            Aedes            aegypti  adult (starved)         female
5            Aedes            aegypti  adult (starved)         female
6            Aedes            aegypti  adult (starved)         female
  Interactor1Temp Interactor1TempUnit Interactor1GrowthTemp
1              22             Celsius                    26
2              22             Celsius                    26
3              22             Celsius                    26
4              22             Celsius                    26
5              22             Celsius                    26
6              22             Celsius                    26
  Interactor1GrowthTempUnit Interactor1Size Interactor1SizeUnit
1                   Celsius            2.93                  mm
2                   Celsius            2.59                  mm
3                   Celsius            2.41                  mm
4                   Celsius            2.62                  mm
5                   Celsius            2.28                  mm
6                   Celsius            2.39                  mm
  Interactor1SizeType            FigureTable
1         wing length supplementary material
2         wing length supplementary material
3         wing length supplementary material
4         wing length supplementary material
5         wing length supplementary material
6         wing length supplementary material
                                                                                                                                  Citation
1 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
2 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
3 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
4 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
5 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
6 Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. R. Soc. B. 288: 20203217.
                     DOI SubmittedBy ContributorEmail Notes
1 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      
2 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      
3 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      
4 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      
5 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      
6 10.1098/rspb.2020.3217 Paul Huxley phuxly@gmail.com      </code></pre>
</div>
</div>
<p>You might want to try the <code>dplyr</code> way to do this data inspection/exploration as well, as you did above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">glimpse</span>(MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 156
Columns: 42
$ Id                        &lt;int&gt; 92846, 92848, 92850, 92852, 92854, 92856, 92…
$ DatasetID                 &lt;int&gt; 558, 558, 558, 558, 558, 558, 558, 558, 558,…
$ IndividualID              &lt;fct&gt; PHX331, PHX332, PHX333, PHX334, PHX335, PHX3…
$ OriginalTraitName         &lt;fct&gt; longevity, longevity, longevity, longevity, …
$ OriginalTraitDef          &lt;fct&gt; individual-level duration of life stage, ind…
$ OriginalTraitValue        &lt;dbl&gt; 8, 5, 5, 6, 4, 6, 9, 1, 7, 7, 7, 5, 9, 7, 7,…
$ OriginalTraitUnit         &lt;fct&gt; days, days, days, days, days, days, days, da…
$ Habitat                   &lt;fct&gt; terrestrial, terrestrial, terrestrial, terre…
$ LabField                  &lt;fct&gt; laboratory, laboratory, laboratory, laborato…
$ SecondStressor            &lt;fct&gt; resource supply, resource supply, resource s…
$ SecondStressorDef         &lt;fct&gt; food quantity supplied at a constant rate , …
$ SecondStressorValue       &lt;dbl&gt; 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1,…
$ SecondStressorUnit        &lt;fct&gt; mg individual-1 day-1, mg individual-1 day-1…
$ LocationText              &lt;fct&gt; Fort Myers Florida, Fort Myers Florida, Fort…
$ LocationType              &lt;fct&gt; field, field, field, field, field, field, fi…
$ CoordinateType            &lt;fct&gt; decimal, decimal, decimal, decimal, decimal,…
$ Latitude                  &lt;dbl&gt; 26.61667, 26.61667, 26.61667, 26.61667, 26.6…
$ Longitude                 &lt;dbl&gt; -81.83333, -81.83333, -81.83333, -81.83333, …
$ Interactor1               &lt;fct&gt; Aedes aegypti, Aedes aegypti, Aedes aegypti,…
$ Interactor1Common         &lt;fct&gt; Yellow fever mosquito, Yellow fever mosquito…
$ Interactor1Kingdom        &lt;fct&gt; Animalia, Animalia, Animalia, Animalia, Anim…
$ Interactor1Phylum         &lt;fct&gt; Arthropoda, Arthropoda, Arthropoda, Arthropo…
$ Interactor1Class          &lt;fct&gt; Insecta, Insecta, Insecta, Insecta, Insecta,…
$ Interactor1Order          &lt;fct&gt; Diptera, Diptera, Diptera, Diptera, Diptera,…
$ Interactor1Family         &lt;fct&gt; Culicidae, Culicidae, Culicidae, Culicidae, …
$ Interactor1Genus          &lt;fct&gt; Aedes, Aedes, Aedes, Aedes, Aedes, Aedes, Ae…
$ Interactor1Species        &lt;fct&gt; aegypti, aegypti, aegypti, aegypti, aegypti,…
$ Interactor1Stage          &lt;fct&gt; adult (starved), adult (starved), adult (sta…
$ Interactor1Sex            &lt;fct&gt; female, female, female, female, female, fema…
$ Interactor1Temp           &lt;dbl&gt; 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, …
$ Interactor1TempUnit       &lt;fct&gt; Celsius, Celsius, Celsius, Celsius, Celsius,…
$ Interactor1GrowthTemp     &lt;dbl&gt; 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, …
$ Interactor1GrowthTempUnit &lt;fct&gt; Celsius, Celsius, Celsius, Celsius, Celsius,…
$ Interactor1Size           &lt;dbl&gt; 2.93, 2.59, 2.41, 2.62, 2.28, 2.39, 2.99, 2.…
$ Interactor1SizeUnit       &lt;fct&gt; mm, mm, mm, mm, mm, mm, mm, mm, mm, mm, mm, …
$ Interactor1SizeType       &lt;fct&gt; wing length, wing length, wing length, wing …
$ FigureTable               &lt;fct&gt; supplementary material, supplementary materi…
$ Citation                  &lt;fct&gt; Huxley et al. 2021. The effect of resource l…
$ DOI                       &lt;fct&gt; 10.1098/rspb.2020.3217, 10.1098/rspb.2020.32…
$ SubmittedBy               &lt;fct&gt; Paul Huxley, Paul Huxley, Paul Huxley, Paul …
$ ContributorEmail          &lt;fct&gt; phuxly@gmail.com, phuxly@gmail.com, phuxly@g…
$ Notes                     &lt;fct&gt; , , , , , , , , , , , , , , , , , , , , , , …</code></pre>
</div>
</div>
<p>To make things clearer, let’s rename some of our columns and remove individuals that don’t have wing lengths (i.e., zeros).</p>
<p>Let’s also define <em>temp</em> and <em>food level</em> as factors because we’ll need this for later.</p>
<p>The <code>tidyverse</code> is particularly useful for this kind of thing!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>MyDF <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(MyDF) <span class="sc">%&gt;%</span> </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rename</span>(<span class="at">winglength =</span> Interactor1Size,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">temp =</span>  Interactor1Temp,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">foodlevel =</span> SecondStressorValue,</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">longevity =</span> OriginalTraitValue) <span class="sc">%&gt;%</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">temp =</span> <span class="fu">factor</span>(temp,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">'22'</span>, <span class="st">'26'</span>, <span class="st">'32'</span>)),</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">foodlevel =</span> <span class="fu">factor</span>(foodlevel,</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">'0.1'</span>, <span class="st">'1'</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(winglength <span class="sc">!=</span> <span class="st">'0'</span>)</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [149 × 42] (S3: tbl_df/tbl/data.frame)
 $ Id                       : int [1:149] 92846 92848 92850 92852 92854 92856 92858 92860 92862 92864 ...
 $ DatasetID                : int [1:149] 558 558 558 558 558 558 558 558 558 558 ...
 $ IndividualID             : Factor w/ 156 levels "PHX331","PHX332",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ OriginalTraitName        : Factor w/ 1 level "longevity": 1 1 1 1 1 1 1 1 1 1 ...
 $ OriginalTraitDef         : Factor w/ 1 level "individual-level duration of life stage": 1 1 1 1 1 1 1 1 1 1 ...
 $ longevity                : num [1:149] 8 5 5 6 4 6 9 1 7 7 ...
 $ OriginalTraitUnit        : Factor w/ 1 level "days": 1 1 1 1 1 1 1 1 1 1 ...
 $ Habitat                  : Factor w/ 1 level "terrestrial": 1 1 1 1 1 1 1 1 1 1 ...
 $ LabField                 : Factor w/ 1 level "laboratory": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressor           : Factor w/ 1 level "resource supply": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressorDef        : Factor w/ 1 level "food quantity supplied at a constant rate ": 1 1 1 1 1 1 1 1 1 1 ...
 $ foodlevel                : Factor w/ 2 levels "0.1","1": 1 1 1 1 1 1 1 1 1 1 ...
 $ SecondStressorUnit       : Factor w/ 1 level "mg individual-1 day-1": 1 1 1 1 1 1 1 1 1 1 ...
 $ LocationText             : Factor w/ 1 level "Fort Myers Florida": 1 1 1 1 1 1 1 1 1 1 ...
 $ LocationType             : Factor w/ 1 level "field": 1 1 1 1 1 1 1 1 1 1 ...
 $ CoordinateType           : Factor w/ 1 level "decimal": 1 1 1 1 1 1 1 1 1 1 ...
 $ Latitude                 : num [1:149] 26.6 26.6 26.6 26.6 26.6 ...
 $ Longitude                : num [1:149] -81.8 -81.8 -81.8 -81.8 -81.8 ...
 $ Interactor1              : Factor w/ 1 level "Aedes aegypti": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Common        : Factor w/ 1 level "Yellow fever mosquito": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Kingdom       : Factor w/ 1 level "Animalia": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Phylum        : Factor w/ 1 level "Arthropoda": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Class         : Factor w/ 1 level "Insecta": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Order         : Factor w/ 1 level "Diptera": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Family        : Factor w/ 1 level "Culicidae": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Genus         : Factor w/ 1 level "Aedes": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Species       : Factor w/ 1 level "aegypti": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Stage         : Factor w/ 1 level "adult (starved)": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1Sex           : Factor w/ 1 level "female": 1 1 1 1 1 1 1 1 1 1 ...
 $ temp                     : Factor w/ 3 levels "22","26","32": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1TempUnit      : Factor w/ 1 level "Celsius": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1GrowthTemp    : num [1:149] 26 26 26 26 26 26 26 26 26 26 ...
 $ Interactor1GrowthTempUnit: Factor w/ 1 level "Celsius": 1 1 1 1 1 1 1 1 1 1 ...
 $ winglength               : num [1:149] 2.93 2.59 2.41 2.62 2.28 2.39 2.99 2.5 2.75 2.82 ...
 $ Interactor1SizeUnit      : Factor w/ 1 level "mm": 1 1 1 1 1 1 1 1 1 1 ...
 $ Interactor1SizeType      : Factor w/ 1 level "wing length": 1 1 1 1 1 1 1 1 1 1 ...
 $ FigureTable              : Factor w/ 1 level "supplementary material": 1 1 1 1 1 1 1 1 1 1 ...
 $ Citation                 : Factor w/ 1 level "Huxley et al. 2021. The effect of resource limitation on the temperature-dependance of mosquito fitness. Proc. "| __truncated__: 1 1 1 1 1 1 1 1 1 1 ...
 $ DOI                      : Factor w/ 1 level "10.1098/rspb.2020.3217": 1 1 1 1 1 1 1 1 1 1 ...
 $ SubmittedBy              : Factor w/ 1 level "Paul Huxley": 1 1 1 1 1 1 1 1 1 1 ...
 $ ContributorEmail         : Factor w/ 1 level "phuxly@gmail.com": 1 1 1 1 1 1 1 1 1 1 ...
 $ Notes                    : Factor w/ 2 levels "","wing length was not measured for this individual: zero assigned ": 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
</div>
</section>
</section>
<section id="scatter-plots" class="level3">
<h3 class="anchored" data-anchor-id="scatter-plots">Scatter Plots</h3>
<p>Let’s start by plotting wing length vs.&nbsp;adult lifespan</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(MyDF<span class="sc">$</span>winglength, MyDF<span class="sc">$</span>longevity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>That doesn’t look very meaningful! Let’s try taking logarithms. <em>Why?</em> - Because taking a log allows you to inspect the data in a meaningful (logarithmic) scale and reveals the true relationship. This also illustrates a important point. Just like statistical analyses, the effectiveness of data visualization too depends on the type of distribution of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log</span>(MyDF<span class="sc">$</span>winglength),<span class="fu">log</span>(MyDF<span class="sc">$</span>longevity))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s look at the same using a base-10 log transform:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log10</span>(MyDF<span class="sc">$</span>longevity),<span class="fu">log10</span>(MyDF<span class="sc">$</span>winglength))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Using a <code>log10</code> transform is often a good idea because then you can see things in terms of “orders of magnitude”, (<span class="math inline">10^1</span>, <span class="math inline">10^2</span>, <span class="math inline">10^3</span>, etc), which makes it easier to determine what the actual values in the original scale are.</p>
<p>Notice also the labels can be changed using <code>xlab</code> and <code>ylab</code>.</p>
<p>We can change almost any aspect of the resulting graph; let’s change the symbols by specifying the <code>plot</code> characters using <code>pch</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log10</span>(MyDF<span class="sc">$</span>longevity),<span class="fu">log10</span>(MyDF<span class="sc">$</span>winglength), <span class="at">xlab =</span> <span class="st">'adult longevity (days)'</span>, <span class="at">ylab =</span> <span class="st">'wing length (mm)'</span>, <span class="at">pch =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A really great summary of basic R graphical parameters can be <a href="https://www.statmethods.net/advgraphs/parameters.html">found here</a>.</p>
</section>
<section id="histograms" class="level3">
<h3 class="anchored" data-anchor-id="histograms">Histograms</h3>
<p>Why did we have to take a logarithm to see the relationship between adult lifespan and wing length? Plotting histograms of the two classes should be insightful, as we can then see the “marginal” distributions of the two variables.</p>
<p>Let’s first plot a histogram of wing length:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(MyDF<span class="sc">$</span>winglength, <span class="at">xlab =</span> <span class="st">'wing length (mm)'</span>, <span class="at">ylab =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s now take a logarithm and see if we can get a better idea of what the distribution of winglengths look like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF<span class="sc">$</span>winglength), <span class="at">xlab =</span> <span class="st">'wing length (mm)'</span>, <span class="at">ylab =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF<span class="sc">$</span>winglength),<span class="at">xlab=</span><span class="st">"log10(wing length (mm))"</span>,<span class="at">ylab=</span><span class="st">"Count"</span>, </span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">"lightblue"</span>, <span class="at">border =</span> <span class="st">"pink"</span>) <span class="co"># Change bar and borders colors</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>So, taking a log really makes clearer what the distribution of wing lengths.</p>
</section>
<section id="subplots" class="level3">
<h3 class="anchored" data-anchor-id="subplots">Subplots</h3>
<p>Lets explore another aspect of this dataset. We will look at the relation of food level to the winglengths. Lets do this by making two subplots next to each other using <code>par</code>. <code>par</code> is a tool to specify the locations of plots relative to each other. The first step is to break the <code>food level</code> aspect into high and low:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>MyDF_lowfood  <span class="ot">&lt;-</span>  MyDF <span class="sc">%&gt;%</span> <span class="fu">filter</span>(foodlevel <span class="sc">==</span> <span class="st">'0.1'</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>MyDF_highfood <span class="ot">&lt;-</span>  MyDF <span class="sc">%&gt;%</span> <span class="fu">filter</span>(foodlevel <span class="sc">==</span> <span class="st">'1'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can start to look at whats going on with it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfcol=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>)) <span class="co"># initialize multi-paneled plot</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="co"># specify which sub-plot to use first </span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF_lowfood<span class="sc">$</span>winglength), <span class="co"># low-histogram</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"log10(wing length (mm))"</span>, <span class="at">ylab =</span> <span class="st">"Count"</span>, <span class="at">col =</span> <span class="st">"lightblue"</span>, <span class="at">border =</span> <span class="st">"pink"</span>, </span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">'0.1 mg/larva/day'</span>) <span class="co"># Add title</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>)) <span class="co"># Second sub-plot</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF_highfood<span class="sc">$</span>winglength), <span class="at">xlab=</span><span class="st">"log10(wing length (mm))"</span>,</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Count"</span>, <span class="at">col =</span> <span class="st">"lightgreen"</span>, <span class="at">border =</span> <span class="st">"pink"</span>, <span class="at">main =</span> <span class="st">'1 mg/larva/day'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The <code>par()</code> function can set multiple graphics parameters (not just multi-panel plots), including figure margins, axis labels, and more. Check out the <a href="https://www.rdocumentation.org/packages/graphics/versions/3.5.1/topics/par">help for this function</a>.</p>
<p>Another option for making multi-panel plots is the <code>layout</code> function.</p>
</section>
<section id="overlaying-plots" class="level3">
<h3 class="anchored" data-anchor-id="overlaying-plots">Overlaying plots</h3>
<p>Better still, we can overlay plots:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="co">#initialize multi-paneled plot</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF_lowfood<span class="sc">$</span>winglength), <span class="co"># high-food histogram</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"log10(wing length (mm))"</span>, <span class="at">ylab=</span><span class="st">"Count"</span>, </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="co"># Note 'rgb', fourth value is transparency</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Larval food-level size overlap"</span>) </span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log10</span>(MyDF_highfood<span class="sc">$</span>winglength), <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), <span class="at">add =</span> T) <span class="co"># Plot high food</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>,<span class="fu">c</span>(<span class="st">'0.1'</span>,<span class="st">'1'</span>),   <span class="co"># Add legend</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill=</span><span class="fu">c</span>(<span class="fu">rgb</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="fu">rgb</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.5</span>))) <span class="co"># Define legend colors</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plot annotation with text can be done with either single or double quotes, i.e., ‘Plot Title’ or “Plot Title”, respectively. But it is generally a good idea to use double quotes because sometimes you would like to use an apostrophe in your title or axis label strings.</p>
</section>
<section id="boxplots" class="level3">
<h3 class="anchored" data-anchor-id="boxplots">Boxplots</h3>
<p>Now, let’s try plotting boxplots instead of histograms. These are useful for getting a visual summary of the distribution of your data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">log10</span>(MyDF<span class="sc">$</span>winglength), <span class="at">xlab =</span> <span class="st">""</span>, <span class="at">ylab =</span> <span class="st">"log10(wing length (mm))"</span>, <span class="at">main =</span> <span class="st">"Wing length"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now let’s see how many temperatures are in included in the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">log</span>(MyDF<span class="sc">$</span>winglength) <span class="sc">~</span> MyDF<span class="sc">$</span>temp<span class="sc">+</span>MyDF<span class="sc">$</span>foodlevel, <span class="co"># Why the tilde?</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"temperature"</span>, <span class="at">ylab =</span> <span class="st">"wing length (mm)"</span>,</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Wing length by temperature"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Note the tilde (<code>~</code>). This is to tell R to subdivide or categorize your analysis and plot by the <code>factor</code> location. More on this later.</p>
<p>To understand this plotting method, think of the full graph area as going from (0,0) in the lower left corner to (1,1) in the upper right corner. The format of the <code>fig=</code> parameter is a numerical vector of the form <code>c(x1, x2, y1, y2)</code>, corresponding to <code>c(bottom, left, top, right)</code>. First, <code>par(fig=c(0,0.8,0,0.8))</code> sets up the scatterplot going from 0 to 0.8 on the x axis and 0 to 0.8 on the y axis, leaving some area for the boxplots at the top and right. The top boxplot goes from 0 to 0.8 on the x axis and 0.4 to 1 on the y axis. The right hand boxplot goes from 0.55 to 1 on the x axis and 0 to 0.8 on the y axis. You can experiment with these proportions to change the spacings between plots.</p>
<p>This plot is useful, because it shows you what the marginal distributions of the two variables are.</p>
<p>Below you will learn to use <code>ggplot</code> to produce a much more elegant, pretty multi-panel plots.</p>
</section>
<section id="saving-your-graphics" class="level3">
<h3 class="anchored" data-anchor-id="saving-your-graphics">Saving your graphics</h3>
<p>And you can also save the figure in a vector graphics format like a pdf. It is important to learn to do this, because you want to be able to save your plots in good resolution, and want to avoid the manual steps of clicking on the figure, doing “save as”, etc. So let’s save the figure as a PDF:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">fig=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.8</span>,<span class="dv">0</span>,<span class="fl">0.8</span>))</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"larvalfoodwing.pdf"</span>, <span class="co"># Open blank pdf page using a relative path</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>    <span class="fl">11.7</span>, <span class="fl">8.3</span>) <span class="co"># These numbers are page dimensions in inches</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log</span>(MyDF_lowfood<span class="sc">$</span>winglength), <span class="co"># Plot predator histogram (note 'rgb')</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"wing length (mm)"</span>, <span class="at">ylab=</span><span class="st">"Count"</span>, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="at">main =</span> <span class="st">"Overlap in wing length by larval food level"</span>) </span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log</span>(MyDF_highfood<span class="sc">$</span>winglength), <span class="co"># Plot prey weights</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), </span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">add =</span> T)  <span class="co"># Add to same plot = TRUE</span></span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'topleft'</span>,<span class="fu">c</span>(<span class="st">'0.1 mg/larva/day'</span>,<span class="st">'1 mg/larva/day'</span>), <span class="co"># Add legend</span></span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill=</span><span class="fu">c</span>(<span class="fu">rgb</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.5</span>), <span class="fu">rgb</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.5</span>))) </span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a><span class="fu">graphics.off</span>() <span class="co"># you can also use dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Always try to save results in a vector format, which can be scaled up to any size. For more on vector vs raster images/graphics, <a href="https://en.wikipedia.org/wiki/Vector_graphics">see this</a>.</p>
<p>Note that you are saving to the <code>results</code> directory now. This is a recommended project organization and workflow: store and retrieve data from a <code>Data</code> directory, keep your code and work from a <code>Code</code> directory, and save outputs to a <code>results</code> directory.</p>
<p>You can also try other graphic output formats. For example, <code>png()</code> (a raster format) instead of <code>pdf()</code>. As always, look at the help documentation of each of these commands!</p>
</section>
</section>
<section id="beautiful-graphics-in-r" class="level2">
<h2 class="anchored" data-anchor-id="beautiful-graphics-in-r">Beautiful graphics in R</h2>
<p>R can produce beautiful visualizations, but it typically takes a lot of work to obtain the desired result. This is because the starting point is pretty much a “bare” plot, and adding features commonly required for publication-grade figures (legends, statistics, regressions, sub-plotting etc.) can require a lot of small and painful additional arguments to the plotting commands <em>at the same time</em>, or even additional steps).</p>
<p>Moreover, it is very difficult to switch from one representation of the data to another (i.e., from boxplots to scatterplots), or to plot several datasets together. The R package <code>ggplot2</code> overcomes these issues, and produces truly high-quality, publication-ready graphics suitable for papers, theses and reports.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>3D plots:</strong> <em>Currently, <code>ggplot2</code> cannot be used to create 3D graphs or mosaic plots (but <a href="https://www.tylermw.com/3d-ggplots-with-rayshader">see this</a>). In any case, most of you won’t be needing 3D plots. If you do, there are many ways to do 3D plots using other plotting packages in R. In particular, look up the <code>scatterplot3d</code> and <code>plot3D</code> packages.</em> <strong><em>I don’t think this is true anymore, I made a weird 3d monster for a paper a while ago.</em></strong></p>
</div>
</div>
</div>
<p><code>ggplot2</code> differs from other approaches as it attempts to provide a “grammar” for graphics in which each layer is the equivalent of a verb, subject etc. and a plot is the equivalent of a sentence. All graphs start with a layer showing the data, other layers and attributes/styles are added to modify the plot. Specifically, according to this grammar, a statistical graphic is a “mapping” from data to geometric objects (points, lines, bars; set using <code>geom</code>) with aesthetic attributes (colour, shape, size; set using <code>aes</code>).</p>
<p>For more on the ideas underlying ggplot, see the book “ggplot2: Elegant Graphics for Data Analysis”, by H. Wickham (in your Reading directory). Also, the <a href="http://ggplot2.tidyverse.org/">ggplot2 website</a> is an excellent resource.</p>
<p>ggplot can be used in two ways: with <code>qplot</code> (for <code>q</code>uick <code>plot</code>ting) and <code>ggplot</code> for fully customized plotting.</p>
<p><em>Note that <code>ggplot2</code> only accepts data in data frames.</em></p>
<section id="quick-plotting-with-qplot" class="level3">
<h3 class="anchored" data-anchor-id="quick-plotting-with-qplot">Quick plotting with <code>qplot</code></h3>
<p><code>qplot</code> can be used to quickly produce graphics for exploratory data analysis, and as a base for more complex graphics. It uses syntax that is closer to the standard R plotting commands.</p>
<p>We will use the same Huxley et al.&nbsp;dataset again – you will soon see how much nicer the same types of plots you made above look when done with ggplot!</p>
<p>First, load the package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>MyDF <span class="ot">&lt;-</span> <span class="fu">subset</span>(MyDF, winglength <span class="sc">!=</span> <span class="st">'NA'</span>) <span class="co"># omit individuals that did not survive to adulthood (i.e. wing length = NA)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="scatterplots" class="level4">
<h4 class="anchored" data-anchor-id="scatterplots">Scatterplots</h4>
<p>Let’s start plotting the <code>wing length</code> vs <code>adult lifespan</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(winglength, longevity, <span class="at">data =</span> MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>As before, let’s take logarithms and plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(<span class="fu">log</span>(winglength), <span class="fu">log</span>(longevity), <span class="at">data =</span> MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now, color the points according to the food level:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(<span class="fu">log</span>(winglength), <span class="fu">log</span>(longevity), <span class="at">col=</span>foodlevel, <span class="at">data =</span> MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>But the figure’s aspect ratio is not very nice. Let’s change it using the <code>asp</code> option:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(<span class="fu">log</span>(winglength), <span class="fu">log</span>(longevity), <span class="at">col=</span>foodlevel, <span class="at">data =</span> MyDF,<span class="at">asp =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The same as above, but changing the shape:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(<span class="fu">log</span>(winglength), <span class="fu">log</span>(longevity), <span class="at">shape=</span>foodlevel, <span class="at">data =</span> MyDF,<span class="at">asp =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="aesthetic-mappings" class="level4">
<h4 class="anchored" data-anchor-id="aesthetic-mappings">Aesthetic mappings</h4>
<p>These examples demonstrate a key difference between <code>qplot</code> (and indeed, ggplot2’s approach) and the standard <code>plot</code> command: When you want to assign colours, sizes or shapes to the points on your plot, using the <code>plot</code> command, it’s your responsibility to convert (i.e., “map”) a categorical variable in your data (e.g., type of feeding interaction in the above case) onto colors (or shapes) that <code>plot</code> knows how to use (e.g., by specifying “red”, “blue”, “green”, etc).</p>
<p>ggplot does this mapping for you automatically, and also provides a legend! This makes it really easy to quickly include additional data (e.g., if a new feeding interaction type was added to the data) on the plot.</p>
<p>Instead of using ggplot’s automatic mapping, if you want to manually set a color or a shape, you have to use <code>I()</code> (meaning “Identity”).</p>
</section>
<section id="setting-transparency" class="level4">
<h4 class="anchored" data-anchor-id="setting-transparency">Setting transparency</h4>
<p>Because there are so many points, we can make them semi-transparent using <code>alpha</code> so that the overlaps can be seen:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(<span class="fu">log</span>(winglength), <span class="fu">log</span>(longevity), <span class="at">col=</span>foodlevel,<span class="at">alpha=</span><span class="fu">I</span>(<span class="fl">0.5</span>), <span class="at">data =</span> MyDF,<span class="at">asp =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here, try using <code>alpha = .5</code> instead of <code>alpha = I(.5)</code> and see what happens.</p>
</section>
</section>
</section>
<section id="box-plots-in-ggplot" class="level2">
<h2 class="anchored" data-anchor-id="box-plots-in-ggplot">Box plots in ggplot</h2>
<p>Boxplots will work largely the same way they have before. We need to make the <code>geom</code> argument in ggplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qplot</span>(temp, <span class="fu">log</span>(winglength), <span class="at">geom =</span> <span class="st">'boxplot'</span>, <span class="at">data =</span> MyDF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="saving-your-plots" class="level4">
<h4 class="anchored" data-anchor-id="saving-your-plots">Saving your plots</h4>
<p>Finally, let’s save a pdf file of the figure (same approach as we <a href="#Saving-your-graphics">used before</a>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"MyFirst-ggplot2-Figure.pdf"</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">qplot</span>(winglength, longevity, <span class="at">data =</span> MyDF,<span class="at">log=</span><span class="st">"xy"</span>,</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">main =</span> <span class="st">"Relationship between wing length and adult lifespan"</span>, </span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">xlab =</span> <span class="st">"log(wing length) (mm)"</span>, </span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">ylab =</span> <span class="st">"log(adult lifespan) (days)"</span>) <span class="sc">+</span> <span class="fu">theme_bw</span>())</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
<p>Using <code>print</code> ensures that the whole command is kept together and that you can use the command in a script.</p>
</section>
<section id="some-more-important-ggplot-options" class="level3">
<h3 class="anchored" data-anchor-id="some-more-important-ggplot-options">Some more important ggplot options</h3>
<p>Other important options to keep in mind:</p>
<table class="table">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<thead>
<tr class="header">
<th>Option</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>xlim</code></td>
<td>limits for x axis: <code>xlim = c(0,12)</code></td>
</tr>
<tr class="even">
<td><code>ylim</code></td>
<td>limits for y axis</td>
</tr>
<tr class="odd">
<td><code>log</code></td>
<td>log transform variable <code>log = "x"</code>, <code>log = "y"</code>, <code>log = "xy"</code></td>
</tr>
<tr class="even">
<td><code>main</code></td>
<td>title of the plot <code>main = "My Graph"</code></td>
</tr>
<tr class="odd">
<td><code>xlab</code></td>
<td>x-axis label</td>
</tr>
<tr class="even">
<td><code>ylab</code></td>
<td>y-axis label</td>
</tr>
<tr class="odd">
<td><code>asp</code></td>
<td>aspect ratio <code>asp = 2</code>, <code>asp = 0.5</code></td>
</tr>
<tr class="even">
<td><code>margins</code></td>
<td>whether or not margins will be displayed</td>
</tr>
</tbody>
</table>
</section>
<section id="the-geom-argument" class="level3">
<h3 class="anchored" data-anchor-id="the-geom-argument">The <code>geom</code> argument</h3>
<p><code>geom</code> Specifies the geometric objects that define the graph type. The geom option is expressed as a R character vector with one or more entries. geom values include “point”, “smooth”, “boxplot”, “line”, “histogram”, “density”, “bar”, and “<a href="https://www.youtube.com/watch?v=3tRFL4q5m7c">jitter</a>”.</p>
</section>
<section id="advanced-plotting-ggplot" class="level3">
<h3 class="anchored" data-anchor-id="advanced-plotting-ggplot">Advanced plotting: <code>ggplot</code></h3>
<p>The command <code>qplot</code> allows you to use only a single dataset and a single set of “aesthetics” (x, y, etc.). To make full use of <code>ggplot2</code>, we need to use the command <code>ggplot</code>, which allows you to use “layering”. Layering is the mechanism by which additional data elements are added to a plot. Each layer can come from a different dataset and have a different aesthetic mapping, allowing us to create plots that could not be generated using <code>qplot()</code>, which permits only a single dataset and a single set of aesthetic mappings.</p>
<p>For a <code>ggplot</code> plotting command, we need at least:</p>
<ul>
<li><p>The data to be plotted, in a data frame;</p></li>
<li><p>Aesthetics mappings, specifying which variables we want to plot, and how;</p></li>
<li><p>The <code>geom</code>, defining the geometry for representing the data;</p></li>
<li><p>(Optionally) some <code>stat</code> that transforms the data or performs statistics using the data.</p></li>
</ul>
<p>To start a graph, we must specify the data and the aesthetics:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(MyDF, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(winglength),</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">y =</span> <span class="fu">log</span>(longevity),</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">colour =</span> foodlevel))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we have created a graphics object <code>p</code> to which we can add layers and other plot elements.</p>
<p>Now try to plot the graph:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plot is blank because we are yet to specify a geometry — only then can we see the graph:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>()</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can use the “+” sign to concatenate different commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fu">I</span>(<span class="dv">2</span>), <span class="at">shape=</span><span class="fu">I</span>(<span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="co"># make the background white</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">aspect.ratio=</span><span class="dv">1</span>) <span class="co">#make the plot square</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s remove the legend:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>q <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">aspect.ratio=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>To make it easier to read, we can plot the smoothed density of the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(MyDF, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(winglength<span class="sc">/</span>longevity), <span class="at">fill =</span> foodlevel )) <span class="sc">+</span> <span class="fu">geom_density</span>()</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.5</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_density</span>()<span class="sc">+</span><span class="fu">facet_wrap</span>( .<span class="sc">~</span> foodlevel)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can also combine categories like this</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">12</span>, <span class="at">repr.plot.height=</span> <span class="dv">14</span>) <span class="co"># Change plot size (in cm)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can also create a multifaceted plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(MyDF, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(winglength), <span class="at">y =</span> <span class="fu">log</span>(longevity))) <span class="sc">+</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>( .<span class="sc">~</span> temp <span class="sc">+</span> foodlevel, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In <code>facet_wrap</code>, You can also free up just the x or y scales; look up the documentation for this function.</p>
</div>
</div>
</div>
</section>
</section>
<section id="mathematical-display" class="level2">
<h2 class="anchored" data-anchor-id="mathematical-display">Mathematical display</h2>
<p>Let’s try mathematical annotation on a axis, and in the plot area.</p>
<p>First create some linear regression “data”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="at">by =</span> <span class="fl">0.1</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">4.</span> <span class="sc">+</span> <span class="fl">0.25</span> <span class="sc">*</span> x <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="fu">length</span>(x), <span class="at">mean =</span> <span class="fl">0.</span>, <span class="at">sd =</span> <span class="fl">2.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and put them in a dataframe</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>my_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>perform a linear regression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>my_lm <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">lm</span>(y <span class="sc">~</span> x, <span class="at">data =</span> my_data))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the data</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(my_data, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y,<span class="at">colour =</span> <span class="fu">abs</span>(my_lm<span class="sc">$</span>residual))) <span class="sc">+</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="fu">expression</span>(alpha<span class="sc">^</span><span class="dv">2</span> <span class="sc">*</span> pi <span class="sc">/</span> beta <span class="sc">*</span> <span class="fu">sqrt</span>(Theta)))</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>add the regression line and throw some math on the plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_abline</span>(<span class="at">intercept =</span> my_lm<span class="sc">$</span>coefficients[<span class="dv">1</span>][<span class="dv">1</span>],</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">slope =</span> my_lm<span class="sc">$</span>coefficients[<span class="dv">2</span>][<span class="dv">1</span>], <span class="at">colour =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">60</span>, <span class="at">y =</span> <span class="dv">0</span>,<span class="at">label =</span> <span class="st">"sqrt(alpha) * 2* pi"</span>), </span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">parse =</span> <span class="cn">TRUE</span>, <span class="at">size =</span> <span class="dv">6</span>, <span class="at">colour =</span> <span class="st">"blue"</span>)</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-61-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Voila! You have just explored complicated data visually!</p>
</section>
<section id="readings-resources" class="level2">
<h2 class="anchored" data-anchor-id="readings-resources">Readings &amp; Resources</h2>
<p>Check out <em>DataDataData!</em>, <em>Visualization</em> and <em>R</em>, under <a href="https://github.com/mhasoba/TheMulQuaBio/tree/master/content/readings"><em>readings</em> on the TheMulQuaBio repository</a>.</p>
<section id="data-management-reformatting-and-cleaning" class="level3">
<h3 class="anchored" data-anchor-id="data-management-reformatting-and-cleaning">Data management, reformatting and cleaning</h3>
<ul>
<li><p>Brian McGill’s <a href="https://dynamicecology.wordpress.com/2016/08/22/ten-commandments-for-good-data-management">Ten commandments for data management</a></p></li>
<li><p>This paper covers similar ground (available in the readings directory): <a href="https://esajournals.onlinelibrary.wiley.com/doi/full/10.1890/0012-9623-90.2.205">Borer et al (2009). Some Simple Guidelines for Effective Data Management. Bulletin of the Ecological Society of America, 90(2), 205-214.</a></p></li>
<li><p><a href="https://onlinelibrary.wiley.com/doi/book/10.1002/0471448354">Dasu, T. &amp; Johnson, T. Exploratory Data Mining and Data Cleaning. (John Wiley &amp; Sons, Inc., 2003). doi:10.1002/0471448354</a></p></li>
<li><p><a href="http://www.theanalysisfactor.com/wide-and-long-data/">Wide vs.&nbsp;long data</a></p></li>
<li><p><a href="http://vita.had.co.nz/papers/tidy-data.pdf">Hadley Wickham’s excellent article about “tidy” data</a>.</p></li>
<li><p><a href="http://vis.stanford.edu/papers/wrangler">Wrangler</a></p></li>
<li><p><a href="https://www2.eecs.berkeley.edu/Pubs/TechRpts/2000/CSD-00-1110.pdf">An interactive framework for data cleaning</a></p></li>
</ul>
</section>
<section id="visualization" class="level3">
<h3 class="anchored" data-anchor-id="visualization">Visualization</h3>
<ul>
<li><p><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003833">Ten Simple Rules for Better Figures</a></p></li>
<li><p>https://clauswilke.com/dataviz/</p></li>
<li><p><a href="https://onlinelibrary.wiley.com/doi/pdf/10.1002/adma.201102518?casa_token=48b0PGtaUiYAAAAA:DQ9bHugikOkHDkf4uLTYIq74negyWefZjAbexYcgYObHn0sks2Q8ulvHcTUWxteqwTb5qAa0IV71Mg">Rolandi et al.&nbsp;“A Brief Guide to Designing Effective Figures for the Scientific Paper”, doi:10.1002/adma.201102518</a></p></li>
<li><p>The classic <a href="https://www.edwardtufte.com/tufte/books_vdqi">Tufte</a>; Available in the Imperial College Central Library. I have also added extracts and a related book in pdf on the git repository. BTW, check out what Tufte thinks of <a href="https://www.edwardtufte.com/tufte/powerpoint">PowerPoint</a>.</p></li>
<li><p><a href="http://motioninsocial.com/tufte">Tufte in R</a></p></li>
</ul>
</section>
</section>
<section id="experimental-design-and-data-exploration" class="level1">
<h1>Experimental design and Data exploration</h1>
<section id="introduction-1" class="level2">
<h2 class="anchored" data-anchor-id="introduction-1">Introduction</h2>
<p>In this section you will learn how to use R to explore your data and determine appropriate statistical analyses.</p>
<p>Ideally, you would like to design experiments (manipulations and/or observations) that are appropriate for the question you want to answer. However, you still need to explore you data to determine what kind of statistical analyses are appropriate for your data, because:</p>
<ol type="a">
<li><p>Your experiments or observations may not go as planned (do they ever?), and</p></li>
<li><p>You might have somebody else’s data to analyse (very common in this era of “Big Data”).</p></li>
</ol>
<p>By the time you have worked through this section, you should be able to:</p>
<ul>
<li><p>Provided sufficient information is available, be able to judge whether the sampling design used to generate a particular dataset was appropriate</p></li>
<li><p>Calculate basic statistical measures on your data to determine its properties.</p></li>
<li><p>Determine if your sample sizes are adequate, especially for a specific statistical test.</p></li>
</ul>
<p>We are going to start off with the simplest of scenarios for statistical testing — that you want to determine whether a sample, or a pair of samples meet some expectation (hypothesis) or not.</p>
<p>First, some let’s revisit some key concepts and terminology.</p>
</section>
<section id="some-statistical-parlance" class="level2">
<h2 class="anchored" data-anchor-id="some-statistical-parlance">Some statistical parlance</h2>
<p>The following terms are important for you to get familiar with:</p>
<ul>
<li><strong>(Statistical) Population</strong>: A <em>complete set</em> of items that share at least one <em>attribute</em> of interest. This attribute of interest is the target of your statistical analysis. For example, if we are interested in studying the weight of year-old cod in the Oceans, the population consists of <em>all</em> year-old cod, but more specifically, the weight measurements of all the individuals of the cod population is what we want to analyse.</li>
</ul>
<hr>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="graphics/collapse-of-Atlantic-cod.jpg" class="img-fluid figure-img" style="width:60.0%"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><p><strong>(Statistical) Distribution</strong>: A mathematical description (expressed as a mathematical equation) of the properties of a population of interest. Theoreticians have come up with a bunch of distributions (e.g., Gaussian or Normal, Poisson, Binomial, etc.) that are appropriate for different kinds of data. Figuring out which distribution best describes a population of interest is one of the first steps in a statistical analysis. The primary goal of experimental design is to collect and measure sufficient individuals of a population to adequately characterize the statistical properties of an attribute (e.g., body weight of yearling cod) of interest. That is, the <em>statistical distribution</em> that best characterizes the attribute.</p></li>
<li><p><strong>(Data or Population) Sample</strong>: A data <em>sample</em> is a set of measurements of the attribute of interest collected from a (statistical) <em>population</em> (all the individuals of interest) by a defined procedure (<em>sampling methodology</em>). In the cod example above, this could be the weight of every individual of a <em>subset</em> (the <em>sample</em>) of the year-old cod population. This could be from a particular location, such as the Atlantic ocean.</p></li>
<li><p><strong>(Statistical) Parameter</strong> : A measure of some attribute of the (theoretical) <em>statistical distribution</em> that is supposed to represent your population. An example would be the average weight of <em>all</em> yearling cod, which presumably follow some sort of distribution. In practice, this is not measurable because the population is much too large or incompletely inaccessible/invisible — imagine measuring the weight of every year-old cod individual in the Atlantic ocean!</p></li>
<li><p><strong>Statistic</strong> (singular): An <em>estimate</em> of a statistical parameter of the population of interest, obtained by calculating the measure for a <em>sample</em>. An example would be the average or mean weight of individuals in a sample of one-year old cod in the Atlantic Ocean. This is also known as a <em>descriptive statistic</em>. Therefore, a <em>Statistic</em> is to a <em>Statistical Parameter</em> what a <em>Sample</em> is to the (Statistical) <em>Population</em>. For example, the average of a sample of cod weights is a statistic that <em>estimates</em> the “real” average of the weights of the entire one-year Cod population (which is its statistical parameter). This real average is also the mean value of the theoretical distribution (e.g., Gaussian) that the population is expected to follow.</p></li>
<li><p><strong>Hypothesis</strong>: An informed <em>postulate</em> about an attribute of your population of interest. For example, you may hypothesize that the one-year old cod population’s mean weight has declined over the last two decades because of preferential fishing of larger individuals. You will typically confront your main hypothesis with a <em>Null Hypothesis</em>, to minimize the risk of making a significant <em>Type I error</em>. This is the probability of wrongly accepting an <em>alternative (or main) hypothesis</em> even is not really true, and rejecting the null hypothesis (e.g., the yearling cods have in reality not declined in weight, but you wrongly infer that they have). This is a big NO NO from a scientific and philosophical standpoint. The rate or probability of the Type I error is denoted by the Greek letter <span class="math inline">\alpha</span>, and equals the <em>significance level</em> of a statistical test. Wrongly rejecting a true alternative (main) hypothesis is also a <em>Type II</em> error.</p></li>
</ul>
</section>
<section id="descriptive-statistics" class="level2">
<h2 class="anchored" data-anchor-id="descriptive-statistics">Descriptive Statistics</h2>
<p>The key statistical measures that describe a sample (or a population) are:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Arithmetic_mean"><strong>Mean</strong> (or average, or more precisely, the arithmetic mean)</a>: This is a measure of the <a href="https://en.wikipedia.org/wiki/Central_tendency"><em>central tendency</em></a> of the sample and population (what values most of the data tend to have). Given a sample <span class="math inline">x_1,x_2,\ldots,x_n</span> of size <span class="math inline">n</span>, the mean is typically denoted by a <span class="math inline">\bar{x}</span>:</li>
</ul>
<p><span class="math display">\bar{x} =  \frac{x_{1} + x_{2} + \dots +x_{n}}{n} = \bar{x} = \frac{\sum_{i=1}^n x_i}{n}</span></p>
<p>That is, it is the sum of all the values in a sample divided by the number, <span class="math inline">n</span>, of items in the sample.</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Standard_deviation"><strong>Standard deviation</strong></a>: A measure of spread of the population around the mean. This is typically denoted by <span class="math inline">s</span>,and is calculated as:</li>
</ul>
<p><span class="math display">s = \sqrt{\frac{(\bar{x} - x_{1})^{2} + (\bar{x} - x_{2})^2 + \dots + (\bar{x} - x_{n})^{2}}{n-1}} =  \sqrt{\frac{\sum_{i=1}^n (\bar{x} - x_{i})^2}{n-1}}</span></p>
<p>That is, it is the square root of the sum of squares (“SS”) of the differences between each item in the sample and the mean, divided by the <em>degrees of freedom, “df”</em> remaining in the data set (<span class="math inline">n-1</span>). df is the sample size, <span class="math inline">n</span>, minus the number of statistical parameters estimated from the data set. This is to reduce the <em>bias</em> in your <em>estimate</em> of the statistic, as you are calculating it from the sample, and not the whole theoretical population.</p>
<p>Thus, the formula for <span class="math inline">s</span> above has <span class="math inline">n-1</span> in its denominator because, to work out the standard deviation, you must have already estimated the mean (<span class="math inline">\bar{x}</span>) from the same data set. This removes 1 degree of freedom. Also, note that the sample variance, <span class="math inline">s^2</span> is the square of standard deviation. Or, in other words, the standard deviation is the square-root of the variance!</p>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Median"><strong>Median</strong></a>: The above two statistics (mean and sd) are particularly meaningful when the sample and population have a symmetric distribution (e.g., normal or gaussian). When the distribution is not symmetric (that is, it is <a href="https://en.wikipedia.org/wiki/Skewness"><em>skewed</em></a>), the <em>median</em> is a better measure of central tendency. This is the middle value in the ordered set of data. That is, exactly 50% of the data lie below and 50% lie above the median.</p></li>
<li><p>Other descriptive statistics you should keep in mind is the <a href="https://en.wikipedia.org/wiki/Range_(statistics)">range</a> (difference between the largest and smallest values), and the <a href="https://en.wikipedia.org/wiki/Quartile">quartiles</a> (values lying in the data divided into the intervals <span class="math inline">[{1\over4}, {1\over 2},{3\over 4}, 1]</span> or at 1% intervals (percentiles). Box-plots, which you have seen, represent a number of these statistics in one figure, as you will soon learn in practice below.</p></li>
</ul>
</section>
<section id="descriptive-statistics-in-r" class="level2">
<h2 class="anchored" data-anchor-id="descriptive-statistics-in-r">Descriptive statistics in R</h2>
<p>Here are the R commands for the key descriptive statistics:</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Command</th>
<th style="text-align: left;">Functionality</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>mean(x)</code></td>
<td style="text-align: left;">Compute mean (of a vector or matrix)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>sd(x)</code></td>
<td style="text-align: left;">Standard deviation</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>var(x)</code></td>
<td style="text-align: left;">Variance</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>median(x)</code></td>
<td style="text-align: left;">Median</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>quantile(x,0.05)</code></td>
<td style="text-align: left;">Compute the 0.05 quantile</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>range(x)</code></td>
<td style="text-align: left;">Range of the data</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>min(x)</code></td>
<td style="text-align: left;">Minimum</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>max(x)</code></td>
<td style="text-align: left;">Maximum</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>sum(x)</code></td>
<td style="text-align: left;">Sum all elements</td>
</tr>
</tbody>
</table>
</section>
<section id="data-types-and-statistical-distributions" class="level2">
<h2 class="anchored" data-anchor-id="data-types-and-statistical-distributions">Data types and (statistical) distributions</h2>
<p>You will typically encounter or sample the following main types of data:</p>
<ul>
<li><p><strong>Continuous numeric</strong>: Data that can take decimal values (real numbers) such as human height or weight. These may be unbounded (any value between negative infinity to positive infinity), or bounded (e.g., between or zero and some upper positive number) like human weight. This is the <code>numeric</code> or <code>real</code> data type in R.</p></li>
<li><p><strong>Discrete numeric</strong>: Integer (whole) numbers such as counts of individuals in a population, e.g., The number of bacteria in a ml of pond water. This is the <code>int</code> data type in R.</p></li>
<li><p><strong>Percentage (proportion)</strong>: A particular kind of numeric data that is strictly bounded between 0 and 100. The fact that you can never get samples of percentages that exceed these bounds makes such data tricky to analyse. This also falls under the <code>numeric</code> or <code>real</code> data type in R.</p></li>
<li><p><strong>Categorical</strong>: Consisting of a fixed number of discrete categories. These are typically stored as a <em>factor</em> in R, with the categories stored as <em>levels</em> in character (string) data type format. For example, the factor “Type.of.feeding.interaction” from the predator-prey dataset you have used previously had five levels: “insectivorous”, “piscivorous”, “planktivorous”, “predacious”, and “predacious/piscivorous”.</p></li>
<li><p><strong>Binary (presence/absence)</strong>: A special type of categorical data are binary, where only two categories/states/values are possible: (1, 0) (or “present”, “absent”) (e.g., a disease symptom). These may be stored as integer, character, or boolean (<code>TRUE</code>/<code>FALSE</code>) in R.</p></li>
</ul>
<p>While designing experiments or exploring data, you need to keep in mind that each data type will often be best-represented by a particular <em>statistical distribution</em>. For example, continuous numeric data are <em>often</em> normally distributed. On the other hand, count data are likely to be distributed according to the Poisson distribution.</p>
<p>If you are lucky, you will mostly have to deal with data that are continuous or discrete numeric, which are the most straightforward to analyse using Linear models (more on that in subsequent sections). However, some of the most interesting and important problems in biology involve proportion (percentage), categorical and binary data (e.g., Presence or absence of a disease symptom).</p>
<p>For example, think about what type of data, and what type of distribution, a sample of the following is likely to be:</p>
<ul>
<li>Wavelength of light</li>
<li>Temperature</li>
<li>Egg clutch size</li>
<li>Rate of a reaction</li>
<li>Eye-colour</li>
<li>Score in Scrabble</li>
<li>Undergraduate Degree class</li>
<li>Ground-cover of grass in a quadrat</li>
<li>Winning side in chess</li>
</ul>
</section>
<section id="sampling-from-distributions-in-r" class="level2">
<h2 class="anchored" data-anchor-id="sampling-from-distributions-in-r">Sampling from distributions in R</h2>
<p>You can generate samples form many statistical distributions in R. This is a handy thing to know because this allows you to simulate a sampling “experiment” . In particular, the following R commands are important:</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Command</th>
<th style="text-align: left;">Functionality</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>rnorm(n, m=0, sd=1)</code></td>
<td style="text-align: left;">Draw <code>n</code> <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally-distributed</a> random numbers with mean = 0 and sd = 1</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>dnorm(x, m=0, sd=1)</code></td>
<td style="text-align: left;">Density function of the normal distribution, which can be used to calculate the probability of a particular event or a set of independent events</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>qnorm(x, m=0, sd=1)</code></td>
<td style="text-align: left;">Cumulative density function</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>runif(n, min=0, max=2)</code></td>
<td style="text-align: left;">Twenty random numbers from uniform distribution with bounds <span class="math inline">[0, 2]</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>rpois(n, lambda=10)</code></td>
<td style="text-align: left;">Twenty random numbers from the Poisson(<span class="math inline">\lambda</span>) distribution</td>
</tr>
</tbody>
</table>
<p>Let’s try some of these. Generate a random sample of 10:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>MySample <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>MySample</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] -0.29599813  0.63782542 -1.82783231  0.94470838  0.04057713 -0.08829885
 [7]  0.02729571 -1.80020822  1.30940039 -0.28104615</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(MySample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-65-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Probability of getting a value of 1 or -1 from a normally distributed random number with mean = 0 and sd = 1:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="dv">1</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2419707</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2419707</code></pre>
</div>
</div>
<p>Probability of getting large values given the same distribution:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="dv">10</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7.694599e-23</code></pre>
</div>
</div>
<p>Very small!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="dv">100</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>Zero!</p>
<p>Look up the documentation and examples for the other commands/functions listed above and many others that are available.</p>
</section>
<section id="two-basic-rules-of-experimental-design-and-sampling" class="level2">
<h2 class="anchored" data-anchor-id="two-basic-rules-of-experimental-design-and-sampling">Two basic rules of experimental design and sampling</h2>
<p>In general, while designing experiments, and sampling from a <em>population</em>, there are two key (and simple) rules:</p>
<ol type="1">
<li><p><strong>The more you sample, the more your sample’s distribution will look like the population distribution</strong> (obviously!)</p></li>
<li><p><strong>The more you sample, the closer will your sample statistic be to the population’s statistical parameter</strong> (the central limit theorem; when the statistical parameter is the mean, this is the “law of large numbers”)</p></li>
</ol>
<p>Let’s test rule 1 using R. We will perform a “experiment” by generating random samples of increasing size from the normal distribution:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>MySample5 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">5</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>) <span class="co"># Draw 5 normal random nos w/ mean 0 and s.d. 1:</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>MySample10 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>) </span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>MySample20 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>) </span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>MySample40 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">40</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>MySample80 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">80</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>MySample160 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">160</span>, <span class="at">m=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s visualize these “samples”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfcol =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)) <span class="co">#initialize multi-paneled plot</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)); <span class="fu">hist</span>(MySample5, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 5'</span>) </span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)); <span class="fu">hist</span>(MySample10, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 10'</span>) </span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)); <span class="fu">hist</span>(MySample20, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 20'</span>) </span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>)); <span class="fu">hist</span>(MySample40, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 40'</span>) </span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)); <span class="fu">hist</span>(MySample80, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 80'</span>) </span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfg =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)); <span class="fu">hist</span>(MySample160, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>), <span class="at">main =</span> <span class="st">'n = 160'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-71-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Rule 2 above states that if I was to repeat even <span class="math inline">n = 5</span> sufficient number of times, you would get a good <em>estimate</em> of mean (= 0) and standard deviation (= 1) of the normal distribution we sampled from. Try doing it. You can take increasing numbers of samples of 5, take their mean, and check how close to get to 0 as you increase your number of samples of 5.</p>
<p>If you give due regard to the two rules of experimental design above, and consider the type of distribution your population of interest follows, you will have taken some basic, important steps towards designing an effective study.</p>
<p>A more rigorous method for designing experiments is to perform a <a href="https://en.wikipedia.org/wiki/Power_(statistics)">power analysis</a>. Power analyses allow you to estimate the minimum sample size required to be able to detect (while minimizing Type I error) an effect of a given size. Covering this is outside the scope of the current course, but you might want to have a look at <a href="https://www.statmethods.net/stats/power.html">this resource</a>.</p>
<p>OK, so you have performed your experiments or field sampling, or some data have fallen into your lap. Let’s move on to data exploration.</p>
</section>
<section id="data-exploration" class="level2">
<h2 class="anchored" data-anchor-id="data-exploration">Data exploration</h2>
<section id="some-general-tips" class="level3">
<h3 class="anchored" data-anchor-id="some-general-tips">Some general tips</h3>
<p>Statistics is a bit of an art.</p>
<p>That is why <em>a priori</em> visualization is important, and directly cutting to the (statistical) chase can often leave one floundering with confusing coefficient estimates and model fits (or over-fits) and overall uncertainty about whether you have picked the right statistical test.</p>
<ul>
<li><p>So no matter what, always first look at the distribution of the response (the “raw data”). If you see multi-modality (multiple peaks), it might mean that some process or effect is generating it. So, in the dragonfly-damselfly example below, a preliminary visualization of the densities of genome size immediately tell you that there are actually two populations (e.g., two levels of effects, or a process that generates the two populations with different central tendencies <span class="math inline">\mu_A</span> and <span class="math inline">\mu_B</span>).</p></li>
<li><p>If it is a regression-type problem, look at the marginal distributions (distributions of the x and y variables) - similar insights can be gained.</p></li>
<li><p>Also, one can look at the distributions within the effects to look for consistency of both shape (sample of the underlying population’s distribution type) and spread (sample of the underlying population’s variance).</p>
<ul>
<li>If there is consistency in both, the battle is pretty much won, because normal or not, there are statistical methods (GLMs, for example) that can be used.</li>
<li>If there is consistency in distribution type but not in variance, half the battle is won. Unequal variances very often tend to be a problem with unbalanced sampling (across effects/treatments), so you can throw in mixed effects (so, GLMMs) which allows you to use empirical information (the data) across effects to help correct for it.</li>
<li>And also, very often, tests are robust to unequal variances (you would need to find a statistical paper, or do your own simulations to support this - so very often better to stick with mixed effects).</li>
<li>If there is consistency in variance but not shape, or inconsistency in both, then you have a harder problem, and may have to resort to transforming some data and not others (not ideal, to be avoided).</li>
</ul></li>
</ul>
</section>
<section id="a-case-study" class="level3">
<h3 class="anchored" data-anchor-id="a-case-study">A case study</h3>
<p>As a case study, we will use data from a paper looking at the relationship between genome size and body size across species of dragonflies and damselflies (<a href="http://en.wikipedia.org/wiki/Odonata">Odonata</a>):</p>
<blockquote class="blockquote">
<p>Ardila-Garcia, AM &amp; Gregory, TR (2009) ‘An exploration of genome size diversity in dragonflies and damselflies (Insecta: Odonata)’ Journal of Zoology, 278, 163 - 173</p>
</blockquote>
<p>You will work with the script file <code>ExpDesign.R</code>, which performs exploratory analyses on the data in <code>GenomeSize.csv</code>. Let’s go through the code block by block.</p>
<p><span class="math inline">\star</span> Get the script <code>ExpDesign.R</code> from the TheMulQuaBio repository and put it in your own <code>Code</code> directory.</p>
<p><span class="math inline">\star</span> Also get <code>GenomeSize.csv</code></p>
<p><span class="math inline">\star</span> Open the script <code>ExpDesign.R</code> in RStudio (or some other text editor).</p>
<p>Use the shift and arrow keys to select the code in block (2), including the comments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>genome <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'activities/data/GenomeSize.csv'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note the relative path <code>../</code>, which will work assuming that you are working from your <code>code</code> directory (that is, you have set your working directory (using <code>setwd()</code>) to <code>code</code>).</p>
<p>This first line (block (1)) reads in the data, as you have learned previously.</p>
<p><span class="math inline">\star</span> Now run the code in block (2) line by line.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genome) <span class="co"># this won't look so nice on your computer!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Suborder    Family              Species GenomeSize GenomeSE GenomeN
1 Anisoptera Aeshnidae    Aeshna canadensis       2.20       NA       1
2 Anisoptera Aeshnidae    Aeshna constricta       1.76     0.06       4
3 Anisoptera Aeshnidae       Aeshna eremita       1.85       NA       1
4 Anisoptera Aeshnidae Aeshna tuberculifera       1.78     0.10       2
5 Anisoptera Aeshnidae       Aeshna umbrosa       2.00       NA       1
6 Anisoptera Aeshnidae    Aeshna verticalis       1.59       NA       1
  BodyWeight TotalLength HeadLength ThoraxLength AdbdomenLength ForewingLength
1      0.159       67.58       6.83        11.81          48.94          45.47
2      0.228       71.97       6.84        10.72          54.41          46.00
3      0.312       78.80       6.27        16.19          56.33          51.24
4      0.218       72.44       6.62        12.53          53.29          49.84
5      0.207       73.05       4.92        11.11          57.03          46.51
6      0.220       66.25       6.48        11.64          48.13          45.91
  HindwingLength ForewingArea HindwingArea MorphologyN
1          45.40       369.57       483.61           2
2          45.48       411.15       517.38           3
3          49.47       460.72       574.33           1
4          48.82       468.74       591.42           2
5          45.97       382.48       481.44           1
6          44.91       400.40       486.97           1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(genome) <span class="co"># Check what the data columns contain</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   100 obs. of  16 variables:
 $ Suborder      : chr  "Anisoptera" "Anisoptera" "Anisoptera" "Anisoptera" ...
 $ Family        : chr  "Aeshnidae" "Aeshnidae" "Aeshnidae" "Aeshnidae" ...
 $ Species       : chr  "Aeshna canadensis" "Aeshna constricta" "Aeshna eremita" "Aeshna tuberculifera" ...
 $ GenomeSize    : num  2.2 1.76 1.85 1.78 2 1.59 1.44 1.16 1.44 1.2 ...
 $ GenomeSE      : num  NA 0.06 NA 0.1 NA NA NA NA NA NA ...
 $ GenomeN       : int  1 4 1 2 1 1 1 1 1 1 ...
 $ BodyWeight    : num  0.159 0.228 0.312 0.218 0.207 0.22 0.344 0.128 0.392 0.029 ...
 $ TotalLength   : num  67.6 72 78.8 72.4 73 ...
 $ HeadLength    : num  6.83 6.84 6.27 6.62 4.92 6.48 7.53 5.74 8.05 5.28 ...
 $ ThoraxLength  : num  11.8 10.7 16.2 12.5 11.1 ...
 $ AdbdomenLength: num  48.9 54.4 56.3 53.3 57 ...
 $ ForewingLength: num  45.5 46 51.2 49.8 46.5 ...
 $ HindwingLength: num  45.4 45.5 49.5 48.8 46 ...
 $ ForewingArea  : num  370 411 461 469 382 ...
 $ HindwingArea  : num  484 517 574 591 481 ...
 $ MorphologyN   : int  2 3 1 2 1 1 4 1 1 1 ...</code></pre>
</div>
</div>
<p>Have a good look at the data. There are three factors (categorical variables): Suborder, splitting the species into dragonflies (Anisoptera) and damselflies (Zygoptera); Family, splitting the species further into 9 taxonomic families; and Species, giving the latin binomial for each species in the table. The remaining columns are measurements of genome size (in picograms) and measurements of body size and morphology (in grams, mm and mm<span class="math inline">^2</span>). There are two columns ending with an N that show the sample size from which the observations for each species are taken and a column ending SE showing standard errors.</p>
<p>One thing you should see in the output from <code>head</code> or <code>str</code> is that there are some observations marked as <code>NA</code> – this is the way R shows <em>missing data</em>. It is important to check how much missing data there are in a dataset, so we’ll use another function that includes this information. Many R functions refuse to use variables containing missing data — this is just R being careful and you can add <code>na.rm=TRUE</code> into most functions to avoid this problem.</p>
<p><span class="math inline">\star</span> Run the <code>summary</code> line from the script window (block 3).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Suborder            Family            Species            GenomeSize    
 Length:100         Length:100         Length:100         Min.   :0.4100  
 Class :character   Class :character   Class :character   1st Qu.:0.7375  
 Mode  :character   Mode  :character   Mode  :character   Median :0.9400  
                                                          Mean   :1.0143  
                                                          3rd Qu.:1.1800  
                                                          Max.   :2.3600  
                                                                          
    GenomeSE          GenomeN        BodyWeight       TotalLength   
 Min.   :0.00600   Min.   : 1.00   Min.   :0.00200   Min.   :22.82  
 1st Qu.:0.02000   1st Qu.: 1.00   1st Qu.:0.01200   1st Qu.:32.35  
 Median :0.03000   Median : 1.00   Median :0.04000   Median :41.41  
 Mean   :0.03398   Mean   : 2.36   Mean   :0.07486   Mean   :43.76  
 3rd Qu.:0.04000   3rd Qu.: 3.00   3rd Qu.:0.10975   3rd Qu.:51.74  
 Max.   :0.10000   Max.   :12.00   Max.   :0.39200   Max.   :82.39  
 NA's   :55                        NA's   :2         NA's   :2      
   HeadLength    ThoraxLength    AdbdomenLength  ForewingLength 
 Min.   :1.11   Min.   : 3.150   Min.   :13.29   Min.   :12.07  
 1st Qu.:1.91   1st Qu.: 5.645   1st Qu.:24.62   1st Qu.:20.76  
 Median :3.75   Median : 7.585   Median :28.39   Median :27.55  
 Mean   :3.62   Mean   : 8.283   Mean   :31.89   Mean   :29.74  
 3rd Qu.:4.93   3rd Qu.:10.720   3rd Qu.:38.12   3rd Qu.:36.89  
 Max.   :8.05   Max.   :16.190   Max.   :61.35   Max.   :55.99  
 NA's   :3      NA's   :2        NA's   :2       NA's   :4      
 HindwingLength   ForewingArea     HindwingArea     MorphologyN    
 Min.   :11.45   Min.   : 21.04   Min.   : 19.39   Min.   : 1.000  
 1st Qu.:19.95   1st Qu.: 62.94   1st Qu.: 56.66   1st Qu.: 1.000  
 Median :26.80   Median :145.95   Median :177.03   Median : 2.000  
 Mean   :28.54   Mean   :174.32   Mean   :205.84   Mean   : 2.949  
 3rd Qu.:35.40   3rd Qu.:226.91   3rd Qu.:279.36   3rd Qu.: 4.000  
 Max.   :54.59   Max.   :527.92   Max.   :632.77   Max.   :13.000  
 NA's   :4       NA's   :4        NA's   :4        NA's   :2       </code></pre>
</div>
</div>
<p>Note that each column gets a separate summary! Look carefully at the output. There is a column for each variable: for factors, it provides a short table of the number of observations in each level and for continuous variables, it provides some simple summary statistics about the distribution (range, quartiles, mean and median), and the number of missing values.</p>
</section>
</section>
<section id="visualise-distributions-of-the-variables" class="level2">
<h2 class="anchored" data-anchor-id="visualise-distributions-of-the-variables">Visualise distributions of the variables</h2>
<p>The <code>summary</code> function shows us the basic distribution (range, quartiles, mean and median) of a continuous variable, but this is easier to interpret if we visualise it. We’ll look at two ways:</p>
<ul>
<li><p><strong>Histogram</strong>: In the simplest form, this shows the number of observations of the variable falling into a set of bins spanning the range of the variable. The option <code>breaks</code> allows you to change the number of bins.</p></li>
<li><p><strong>Density plot</strong>: Rather than showing blocks of counts, the density plot shows a continuous smooth line. This is a <em>smoothed</em> estimate of the how frequently data is observed across the range of values and the <em>bandwidth</em> (<code>bw=0.1</code>) controls the degree of the smoothing.</p></li>
</ul>
<p><span class="math inline">\star</span> Go to block (4) of the script and run each line separately, looking at the output.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(genome<span class="sc">$</span>GenomeSize, <span class="at">breaks=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(genome<span class="sc">$</span>GenomeSize, <span class="at">bw=</span><span class="fl">0.1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-77-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In your code editor, change the values of <code>breaks</code> and <code>bw</code> (gfor example <code>breaks=5</code> and <code>bw=0.05</code>), and re-run these lines to see how this affects the graph. Basically, with both types of graph you can look at the data too coarsely or too finely.</p>
<p>The graphs you’ve just created look at genome size. Add a copy of those two lines of code in the script and change them to look at the variable <code>TotalLength</code>. You will need to alter the <code>density</code> function to ignore missing values (<code>na.rm=TRUE</code>) and to play around with the bandwidth. You should get something like this:</p>
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="graphics/histTL1.png" class="img-fluid figure-img" style="width:40.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="graphics/histTL2.png" class="img-fluid figure-img" style="width:40.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="take-a-quick-look-at-effects-of-certain-factors" class="level2">
<h2 class="anchored" data-anchor-id="take-a-quick-look-at-effects-of-certain-factors">Take a quick look at effects of certain factors</h2>
<p>R has a special way of describing a model that defines the response variable and the explanatory variables (“factors”). This is called a ‘formula’ and is used to define linear models (more on these in a later sections). The same structure is used in many plotting functions and will put the response variable on the <span class="math inline">y</span> axis and the explanatory variable on the <span class="math inline">x</span> axis. The structure is “response variable ~ explanatory variables”. We will look at multiple explanatory variables in a later section but an example with one explantory variable (factor) is:</p>
<p><code>Genome Size ~ Suborder</code></p>
<p>This formula tells R to model genome size ‘as a function of’ (<code>~</code>) the suborders of Odonata. When using this syntax in a <code>plot</code> function, the result will be to plot genome size as a function of the suborders.</p>
</section>
<section id="compare-distribution-of-the-variable-across-levels-of-a-factor" class="level2">
<h2 class="anchored" data-anchor-id="compare-distribution-of-the-variable-across-levels-of-a-factor">Compare distribution of the variable across levels of a factor</h2>
<p>Although looking at the distribution of variables is a good first step, we often want to compare distributions. In this case, we might want to know how genome size varies between dragonflies and damselflies. The first way we will look at is using boxplots — these show the median and the 25% and 75% quantiles as a box, with whiskers extending to the minimum and maximum. More extreme outliers are plotted independently as points. The <code>plot</code> function in R automatically generates a boxplot when the explanatory variable is a factor.</p>
<p><span class="math inline">\star</span> Go to block 5 of the script and run the first line, looking at genome size between the two suborders:</p>
<p>Duplicate and alter this line to look at the same plot for total length. You should get a plot like this:</p>
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="graphics/bxpTL.png" class="img-fluid figure-img" style="width:40.0%"></p>
</figure>
</div>
</div>
</div>
<p>Although histograms are great for one variable, plotting two histograms on top of one another rarely works well because the overlapping bars are hard to interpret (recall the predator-prey body size example). Density plots don’t have this problem, but it takes a bit more code to create the plot.</p>
<p><span class="math inline">\star</span> Block 6 of the script uses the <code>subset</code> function to create two new data frames separating the data for dragonflies and damselflies. Run the first two lines of this block:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>Anisoptera <span class="ot">&lt;-</span> <span class="fu">subset</span>(genome, Suborder<span class="sc">==</span><span class="st">'Anisoptera'</span>) <span class="co">#The dragonflies</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>Zygoptera <span class="ot">&lt;-</span> <span class="fu">subset</span>(genome, Suborder<span class="sc">==</span><span class="st">'Zygoptera'</span>) <span class="co">#The damselflies</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remember that the arrow symbol (<code>&lt;-</code>) is used to save the output of a function into a new object in R — if you use <code>ls()</code> in the console, you will see the two new data frames.</p>
<p>In the console, use <code>str</code> and <code>summary</code> to explore these two new dataframes. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Anisoptera)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Suborder    Family              Species GenomeSize GenomeSE GenomeN
1 Anisoptera Aeshnidae    Aeshna canadensis       2.20       NA       1
2 Anisoptera Aeshnidae    Aeshna constricta       1.76     0.06       4
3 Anisoptera Aeshnidae       Aeshna eremita       1.85       NA       1
4 Anisoptera Aeshnidae Aeshna tuberculifera       1.78     0.10       2
5 Anisoptera Aeshnidae       Aeshna umbrosa       2.00       NA       1
6 Anisoptera Aeshnidae    Aeshna verticalis       1.59       NA       1
  BodyWeight TotalLength HeadLength ThoraxLength AdbdomenLength ForewingLength
1      0.159       67.58       6.83        11.81          48.94          45.47
2      0.228       71.97       6.84        10.72          54.41          46.00
3      0.312       78.80       6.27        16.19          56.33          51.24
4      0.218       72.44       6.62        12.53          53.29          49.84
5      0.207       73.05       4.92        11.11          57.03          46.51
6      0.220       66.25       6.48        11.64          48.13          45.91
  HindwingLength ForewingArea HindwingArea MorphologyN
1          45.40       369.57       483.61           2
2          45.48       411.15       517.38           3
3          49.47       460.72       574.33           1
4          48.82       468.74       591.42           2
5          45.97       382.48       481.44           1
6          44.91       400.40       486.97           1</code></pre>
</div>
</div>
<p>Now that we’ve got the data separated we can go about plotting the two curves.</p>
<p><span class="math inline">\star</span> Run the next two lines of code in block 6. The first draws the plot for damselflies and the second adds a line for the dragonflies:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(Zygoptera<span class="sc">$</span>GenomeSize), <span class="at">xlim=</span><span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">2.7</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.7</span>))</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">density</span>(Anisoptera<span class="sc">$</span>GenomeSize), <span class="at">col=</span><span class="st">'red'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-82-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Duplicate these last two lines of code and edit them to generate a similar plot for total body length. You will need to edit the code to change the range of the <span class="math inline">x</span> and <span class="math inline">y</span> axes (<code>xlim</code> and <code>ylim</code>) to get both curves to fit neatly on to the graph. It should look like this:</p>
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="graphics/densTL.png" class="img-fluid figure-img" style="width:40.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="explore-further-by-scatter-plotting-two-variables" class="level2">
<h2 class="anchored" data-anchor-id="explore-further-by-scatter-plotting-two-variables">Explore further by scatter-plotting two variables</h2>
<p>Once we’ve looked at the distribution of variables, the next thing is to look at the relationships between continuous variables using scatterplots. The <code>plot</code> function in R automatically generates a scatterplot when the explanatory variable is continuous, so we can use the same syntax and structure as for the boxplot above.</p>
<p><span class="math inline">\star</span> Go to block (7) of the script and run the plot commands. The third one plots genome size as a function of body weight:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GenomeSize <span class="sc">~</span> BodyWeight, <span class="at">data =</span> genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="data_wrangling_files/figure-html/unnamed-chunk-84-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The scatterplot seems to show a weak relationship between genome size and morphology. But maybe dragonflies and damselflies show different relationships, and we can’t distinguish between them! To explore this possibility, we need to plot the two orders using different colours or plot characters. In the next code block, we will customize the plots to show different types of points for each suborder. It is done by using <em>indexing</em>.</p>
<p><span class="math inline">\star</span> Run the first three lines of code in block 8. There are two levels of suborder and these two lines set up a colour and a plot symbol that will be used for each one.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(genome<span class="sc">$</span>Suborder) <span class="co">#Confirm that there are two levels under suborders </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> chr [1:100] "Anisoptera" "Anisoptera" "Anisoptera" "Anisoptera" ...</code></pre>
</div>
</div>
<p>You can see that there are two levels, with Anisoptera first and then Zygoptera. You can also see that these are stored as numeric values: 1 refers to the first level and 2 the second. We can use these as <em>indices</em> to pair the colours and plot symbols to each suborder. These are set in the <code>plot</code> function using the options <code>col=</code> and <code>pch=</code>, which stands for ” <code>p</code>lot <code>ch</code>aracter”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>myColours <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'red'</span>, <span class="st">'blue'</span>) <span class="co"># So choose two colours</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>mySymbols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>) <span class="co"># And two different markers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Run the next plot command to see the resulting plot:</p>
<p>Thus each point gets the appropriate colour and symbol for its group. This is the indexing: <code>myColours[Suborder]</code> and <code>mySymbols[Suborder]</code> automatically assign the two colors and two symbols to the two suborders.</p>
<p>There are a lot of built in colours and plot symbols in R, so the next thing to experiment with is changing these to your own versions.</p>
<p><span class="math inline">\star</span> In the R commandline, type in the function <code>colors()</code>.</p>
<p>You’ll see a long list of options to choose from, so pick two to replace red and blue in the plot above.</p>
<p>The options for the plot symbols are shown below. Pick two to replace the current symbol choices.</p>
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="graphics/pch.png" class="img-fluid figure-img" style="width:40.0%"></p>
</figure>
</div>
</div>
</div>
<p>Rerun the <code>plot</code> function and see what you get!</p>
</section>
<section id="save-your-results" class="level2">
<h2 class="anchored" data-anchor-id="save-your-results">Save your results</h2>
<section id="saving-the-exploratory-graphics" class="level3">
<h3 class="anchored" data-anchor-id="saving-the-exploratory-graphics">Saving the exploratory graphics</h3>
<p>The function <code>pdf</code> opens a new empty pdf file which can then be used to plot graphs. You can set the width and the height of the page size in the pdf but note that this is set in <em>inches</em>. When you have finished creating a plot, the function <code>dev.off</code> closes the pdf file and makes it readable.</p>
<p><span class="math inline">\star</span> Open ‘GenomeSize.pdf’ in a PDF reader. It uses the original colours and plot symbols. Close the file and then delete it from the folder.</p>
<p><span class="math inline">\star</span> Now go back to the script in R and select and run all the code in block (9)</p>
<p>Go back to the <code>results</code> folder. The pdf file should have been recreated — open it and it should now use your choice of colours and symbols.</p>
</section>
<section id="saving-other-results" class="level3">
<h3 class="anchored" data-anchor-id="saving-other-results">Saving other results</h3>
<p>You can also save the data and variables in R format — the original data, two subsets of the data and the two sets of colours and symbols. This will make it easy to restart where you left off. However, We can recreate the data subsets easily, so we’ll just save the data and your colour sets.</p>
<p><span class="math inline">\star</span> Go to the script window and run the final line in block (10)</p>
<p>Still in the script window, choose ‘File <span class="math inline">\triangleright</span> Save’ to save your changes to the script file.</p>
<p>Quit from R by typing <code>q()</code> at the R command prompt. You can also use ctrl+D in Unix).</p>
</section>
</section>
<section id="from-data-exploration-to-statistical-analysis" class="level2">
<h2 class="anchored" data-anchor-id="from-data-exploration-to-statistical-analysis">From data exploration to statistical analysis</h2>
<p>After you have performed your data exploration, you are in a position to make an informed decision about what statistical analyses to perform. Here is a decision tree that you can use, and which includes the methods you will learn in the following sections:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="graphics/yr1_2_flowchart.svg" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="practice-problem-for-data-wrangling" class="level2">
<h2 class="anchored" data-anchor-id="practice-problem-for-data-wrangling">Practice Problem for Data Wrangling</h2>
<p>Organise (or wrangle!) the data, as you learned in the lesson. The dataset provided <code>wranglingdataset.csv</code> contains values on trait-temperature relationships in three important vector/pest species. Develop a hypothesis (or a question) related to how you expect a trait(s) responds to temperature and explore it. Present your group’s findings using descriptive statistics and visualizations.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>